{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "61187df2",
   "metadata": {},
   "source": [
    "# Compute and Analyze Classifier Metrics After Controlling for GFP\n",
    "\n",
    "The snakemake pipeline output the probability of 0 / 1 for each cell for each classifier. Here, we compute and save many common metrics from these probabilities. Here we analyze the classification results of each allele."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "2b5eb95d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "import os\n",
    "import glob\n",
    "import operator\n",
    "import polars as pl\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "from pathlib import Path\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from collections import defaultdict\n",
    "from functools import reduce\n",
    "import sys\n",
    "sys.path.append(\"../..\")\n",
    "from img_utils import *\n",
    "\n",
    "letter_dict_rev = {v: k for k, v in letter_dict.items()}\n",
    "channel_dict_rev = {v: k for k, v in channel_dict.items()}\n",
    "channel_list = list(channel_dict_rev.values())[:-3]\n",
    "\n",
    "%matplotlib inline\n",
    "## Disable truncation globally\n",
    "# pl.Config.set_tbl_rows(20)  # Show all rows\n",
    "# pl.Config.set_tbl_cols(40)  # Show all columns\n",
    "\n",
    "## define control types\n",
    "TC = [\"EGFP\"]\n",
    "NC = [\"RHEB\", \"MAPK9\", \"PRKACB\", \"SLIRP\"]\n",
    "PC = [\"ALK\", \"ALK_Arg1275Gln\", \"PTK2B\"]\n",
    "cNC = [\"Renilla\"]\n",
    "cPC = [\"KRAS\", \"PTK2B\", \"GHSR\", \"ABL1\", \"BRD4\", \"OPRM1\", \"RB1\", \"ADA\", \"WT PMP22\", \"LYN\", \"TNF\", \n",
    "       \"CYP2A6\", \"CSK\", \"PAK1\", \"ALDH2\", \"CHRM3\", \"KCNQ2\", \"ALK T1151M\", \"PRKCE\", \"LPAR1\", \"PLP1\"]\n",
    "\n",
    "TRN_IMBAL_THRES = 3\n",
    "MIN_CLASS_NUM = 2\n",
    "AUROC_THRESHOLDS = [0.95] ## , 0.99"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "60b23c5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "allele_meta_df = pl.DataFrame()\n",
    "allele_meta_df_dict, img_well_qc_sum_dict = {}, {}\n",
    "\n",
    "for bio_rep, bio_rep_batches in BIO_REP_BATCHES_DICT.items():\n",
    "    for batch_id in bio_rep_batches:\n",
    "        allele_meta_df_batch = pl.DataFrame()\n",
    "        platemaps = [file for file in os.listdir(PLATEMAP_DIR.format(batch_id=batch_id)) if file.endswith(\".txt\")]\n",
    "        for platemap in platemaps:\n",
    "            platemap_df = pl.read_csv(os.path.join(PLATEMAP_DIR.format(batch_id=batch_id), platemap), separator=\"\\t\", infer_schema_length=100000)\n",
    "            allele_meta_df_batch = pl.concat([allele_meta_df_batch, \n",
    "                                        platemap_df.filter((pl.col(\"node_type\").is_not_null()))], # (~pl.col(\"node_type\").is_in([\"TC\",\"NC\",\"PC\"]))&\n",
    "                                        how=\"diagonal_relaxed\").sort(\"plate_map_name\")\n",
    "            allele_meta_df_batch = allele_meta_df_batch.with_columns(pl.col(\"plate_map_name\").alias(\"plate_map\")) ## str.split('_').list.get(0).\n",
    "            # display(allele_meta_df.head())\n",
    "        allele_meta_df = pl.concat([\n",
    "            allele_meta_df,\n",
    "            allele_meta_df_batch\n",
    "        ], how=\"diagonal_relaxed\")#.sort(\"plate_map_name\") ## (~pl.col(\"node_type\").is_in([\"TC\",\"NC\",\"PC\"]))&\n",
    "    allele_meta_df_dict[bio_rep] = allele_meta_df_batch\n",
    "\n",
    "    img_well_qc_sum = pl.read_csv(f\"{IMGS_QC_METRICS_DIR}/{bio_rep}/plate-well-level_img_qc_sum.csv\")\n",
    "    img_well_qc_sum = img_well_qc_sum.with_columns(\n",
    "        pl.col(\"channel\").replace(\"DAPI\", \"DNA\").alias(\"channel\")\n",
    "    )\n",
    "    img_well_qc_sum_morph = img_well_qc_sum.filter(pl.col(\"channel\")!=\"GFP\")\n",
    "    img_well_qc_sum_morph = img_well_qc_sum_morph.group_by([\"plate\",\"well\"]).agg(\n",
    "        pl.col(\"is_bg\").max().alias(\"is_bg\"),\n",
    "        pl.col(\"s2n_ratio\").mean().alias(\"s2n_ratio\")\n",
    "    ).with_columns(pl.lit(\"Morph\").alias(\"channel\"))\n",
    "    img_well_qc_sum = pl.concat([\n",
    "        img_well_qc_sum.select(pl.col([\"plate\",\"well\",\"channel\",\"is_bg\",\"s2n_ratio\"])),\n",
    "        img_well_qc_sum_morph.select(pl.col([\"plate\",\"well\",\"channel\",\"is_bg\",\"s2n_ratio\"])),\n",
    "    ], how=\"vertical_relaxed\")\n",
    "    img_well_qc_sum_dict[bio_rep] = img_well_qc_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "65d6fbdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "class_info_dir = \"{}/{}/profiles_tcdropped_filtered_var_mad_outlier_featselect_filtcells/metrics_gfp_adj.csv\"\n",
    "\n",
    "for bio_rep, bio_rep_batches in BIO_REP_BATCHES_DICT.items():\n",
    "    for batch_id in bio_rep_batches:\n",
    "        if bio_rep != \"2024_12_Batch_11-12\":\n",
    "            batch_id_tmp = batch_id + \"_morph\"\n",
    "        class_res_df = pl.read_csv(class_info_dir.format(CLASS_ANALYSES_DIR, batch_id_tmp))\n",
    "        class_res_df = class_res_df.with_columns(\n",
    "            pl.when((pl.col(\"well_0\").str.contains(r\"(?:01|P|A|24)\"))|(pl.col(\"well_1\").str.contains(r\"(?:01|P|A|24)\")))\n",
    "            .then(pl.lit(True))\n",
    "            .otherwise(pl.lit(False))\n",
    "            .alias(\"Well_On_Edge\")\n",
    "        )\n",
    "        # display(class_res_df)\n",
    "        for channel in [\"GFP\"]:\n",
    "            class_res_df_channel = class_res_df.filter(pl.col(\"Metadata_Feature_Type\")==channel)\n",
    "            class_res_df_channel = class_res_df_channel.join(\n",
    "                img_well_qc_sum.filter(pl.col(\"channel\")==channel).select(pl.col([\"plate\",\"well\",\"is_bg\"])),\n",
    "                left_on=[\"Plate\",\"well_0\"],\n",
    "                right_on=[\"plate\",\"well\"]\n",
    "            ).rename({\"is_bg\": \"well_0_is_bg\"})\n",
    "            class_res_df_channel = class_res_df_channel.join(\n",
    "                img_well_qc_sum.filter(pl.col(\"channel\")==channel).select(pl.col([\"plate\",\"well\",\"is_bg\"])),\n",
    "                left_on=[\"Plate\",\"well_1\"],\n",
    "                right_on=[\"plate\",\"well\"]\n",
    "            ).rename({\"is_bg\": \"well_1_is_bg\"})\n",
    "            \n",
    "            plate_maps = sorted(set([\"_\".join(pm.split(\"_\")[:-1]) for pm in class_res_df_channel[\"Plate\"].unique()]))\n",
    "            for pm in plate_maps:\n",
    "                class_res_ch_pm = class_res_df_channel.filter(pl.col(\"Plate\").str.contains(pm))\n",
    "                plates = sorted(class_res_ch_pm[\"Plate\"].unique().to_list())\n",
    "                # fig, axes = plt.subplots(2,2,figsize=(48,23)) ## sharey=True,sharex=True\n",
    "                for plate in plates:\n",
    "                    plate_info = class_res_ch_pm.filter(pl.col(\"Plate\")==plate)\n",
    "    \n",
    "                    ## plot the ctrls alleles first\n",
    "                    plate_info_ctrl = plate_info.filter(pl.col(\"Metadata_Control\"))\n",
    "                    ctrls_wells = pl.concat([plate_info_ctrl[\"well_0\"], plate_info_ctrl[\"well_1\"]]).unique()\n",
    "                    agg_group_by_well_0 = plate_info_ctrl.group_by(\"well_0\",\"allele_0\").agg([\n",
    "                        pl.col(\"AUROC\").mean().alias(\"AUROC_Mean\"),\n",
    "                        pl.col(\"well_0_is_bg\").max().alias(\"is_bg\")\n",
    "                    ]).rename(\n",
    "                        {\"well_0\": \"well\", \"allele_0\": \"allele\"}\n",
    "                    )#.with_columns(pl.col(\"Well_On_Edge\").cast(pl.Boolean).alias(\"Well_On_Edge\"))\n",
    "                    agg_group_by_well_1 = plate_info_ctrl.filter(\n",
    "                        (pl.col(\"well_1\").is_in(ctrls_wells))&\\\n",
    "                        (~pl.col(\"well_1\").is_in(agg_group_by_well_0[\"well\"]))\n",
    "                    ).group_by(\"well_1\",\"allele_1\").agg([\n",
    "                        pl.col(\"AUROC\").mean().alias(\"AUROC_Mean\"),\n",
    "                        pl.col(\"well_1_is_bg\").max().alias(\"is_bg\")\n",
    "                    ]).rename(\n",
    "                        {\"well_1\": \"well\", \"allele_1\": \"allele\"}\n",
    "                    )\n",
    "                    agg_group_by_well = pl.concat(\n",
    "                        [\n",
    "                            agg_group_by_well_0.select(pl.col(\"well\",\"AUROC_Mean\",\"allele\",\"is_bg\")),\n",
    "                            agg_group_by_well_1.select(pl.col(\"well\",\"AUROC_Mean\",\"allele\",\"is_bg\"))\n",
    "                        ]\n",
    "                    )\n",
    "                    ## plot the auroc per each plate for a platemap\n",
    "                #     plot_platemap(\n",
    "                #         agg_group_by_well,\n",
    "                #         plate+f\"_{channel}\",\n",
    "                #         well_pos_col=\"well\",\n",
    "                #         # this is the column to color by (categorical or continuous)\n",
    "                #         value_col=\"AUROC_Mean\",\n",
    "                #         # these columns will be concatenated into the annotation text\n",
    "                #         label_cols=(\"allele\",\"AUROC_Mean\"),\n",
    "                #         ax=axes[plates.index(plate)//2, plates.index(plate)%2],\n",
    "                #         value_type=\"continuous\",   # or \"continuous\"\n",
    "                #         continuous_cmap=\"vlag\",  # matplotlib colormap for continuous mode\n",
    "                #         categorical_colors={True: \"tomato\", False: \"skyblue\"},     # dict for categorical → color\n",
    "                #         grid_square=\"is_bg\"\n",
    "                #     )\n",
    "                # fig.subplots_adjust(wspace=-.55, hspace=.05)\n",
    "                # plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "5ce6c102",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=====================================B7 metrics=====================================\n",
      "==========================================================================\n",
      "Total number of unique classifiers for GFP: 372\n",
      "Total number of unique variant alleles: 372\n",
      "Total number of unique WT genes: 77\n",
      "==========================================================================\n",
      "After filtering out classifiers with training imbalance > 3:\n",
      "Total number of unique classifiers: (375, 4)\n",
      "Total number of unique variant alleles: 375\n",
      "Total number of unique WT genes: 77\n",
      "==========================================================================\n",
      "=====================================B8 metrics=====================================\n",
      "==========================================================================\n",
      "Total number of unique classifiers for GFP: 259\n",
      "Total number of unique variant alleles: 259\n",
      "Total number of unique WT genes: 62\n",
      "==========================================================================\n",
      "After filtering out classifiers with training imbalance > 3:\n",
      "Total number of unique classifiers: (263, 4)\n",
      "Total number of unique variant alleles: 263\n",
      "Total number of unique WT genes: 65\n",
      "==========================================================================\n",
      "=====================================B11 metrics=====================================\n",
      "==========================================================================\n",
      "Total number of unique classifiers for GFP: 0\n",
      "Total number of unique variant alleles: 0\n",
      "Total number of unique WT genes: 0\n",
      "==========================================================================\n",
      "After filtering out classifiers with training imbalance > 3:\n",
      "Total number of unique classifiers: (0, 3)\n",
      "Total number of unique variant alleles: 0\n",
      "Total number of unique WT genes: 0\n",
      "==========================================================================\n",
      "=====================================B12 metrics=====================================\n",
      "==========================================================================\n",
      "Total number of unique classifiers for GFP: 0\n",
      "Total number of unique variant alleles: 0\n",
      "Total number of unique WT genes: 0\n",
      "==========================================================================\n",
      "After filtering out classifiers with training imbalance > 3:\n",
      "Total number of unique classifiers: (0, 3)\n",
      "Total number of unique variant alleles: 0\n",
      "Total number of unique WT genes: 0\n",
      "==========================================================================\n",
      "=====================================B15 metrics=====================================\n",
      "==========================================================================\n",
      "Total number of unique classifiers for GFP: 223\n",
      "Total number of unique variant alleles: 223\n",
      "Total number of unique WT genes: 34\n",
      "==========================================================================\n",
      "After filtering out classifiers with training imbalance > 3:\n",
      "Total number of unique classifiers: (223, 4)\n",
      "Total number of unique variant alleles: 223\n",
      "Total number of unique WT genes: 34\n",
      "==========================================================================\n",
      "=====================================B16 metrics=====================================\n",
      "==========================================================================\n",
      "Total number of unique classifiers for GFP: 232\n",
      "Total number of unique variant alleles: 232\n",
      "Total number of unique WT genes: 34\n",
      "==========================================================================\n",
      "After filtering out classifiers with training imbalance > 3:\n",
      "Total number of unique classifiers: (232, 4)\n",
      "Total number of unique variant alleles: 232\n",
      "Total number of unique WT genes: 34\n",
      "==========================================================================\n",
      "=====================================B13 metrics=====================================\n",
      "==========================================================================\n",
      "Total number of unique classifiers for GFP: 464\n",
      "Total number of unique variant alleles: 464\n",
      "Total number of unique WT genes: 26\n",
      "==========================================================================\n",
      "After filtering out classifiers with training imbalance > 3:\n",
      "Total number of unique classifiers: (464, 4)\n",
      "Total number of unique variant alleles: 464\n",
      "Total number of unique WT genes: 26\n",
      "==========================================================================\n",
      "=====================================B14 metrics=====================================\n",
      "==========================================================================\n",
      "Total number of unique classifiers for GFP: 458\n",
      "Total number of unique variant alleles: 458\n",
      "Total number of unique WT genes: 26\n",
      "==========================================================================\n",
      "After filtering out classifiers with training imbalance > 3:\n",
      "Total number of unique classifiers: (458, 4)\n",
      "Total number of unique variant alleles: 458\n",
      "Total number of unique WT genes: 26\n",
      "==========================================================================\n",
      "=====================================B18 metrics=====================================\n",
      "==========================================================================\n",
      "Total number of unique classifiers for GFP: 286\n",
      "Total number of unique variant alleles: 286\n",
      "Total number of unique WT genes: 42\n",
      "==========================================================================\n",
      "After filtering out classifiers with training imbalance > 3:\n",
      "Total number of unique classifiers: (307, 4)\n",
      "Total number of unique variant alleles: 307\n",
      "Total number of unique WT genes: 45\n",
      "==========================================================================\n",
      "=====================================B19 metrics=====================================\n",
      "==========================================================================\n",
      "Total number of unique classifiers for GFP: 296\n",
      "Total number of unique variant alleles: 296\n",
      "Total number of unique WT genes: 44\n",
      "==========================================================================\n",
      "After filtering out classifiers with training imbalance > 3:\n",
      "Total number of unique classifiers: (312, 4)\n",
      "Total number of unique variant alleles: 312\n",
      "Total number of unique WT genes: 47\n",
      "==========================================================================\n"
     ]
    }
   ],
   "source": [
    "metric_df_dict = {}\n",
    "\n",
    "for bio_rep, bio_rep_batches in BIO_REP_BATCHES_DICT.items():\n",
    "    for batch_id in bio_rep_batches:\n",
    "        if bio_rep != \"2024_12_Batch_11-12\":\n",
    "            batch_id_tmp = batch_id + \"_morph\"\n",
    "            \n",
    "        metrics_df = pl.read_csv(class_info_dir.format(CLASS_ANALYSES_DIR, batch_id_tmp))\n",
    "        batch_id = f\"B{batch_id.split('Batch_')[-1]}\"\n",
    "        # display(metrics_df)\n",
    "        print(f\"====================================={batch_id} metrics=====================================\")\n",
    "        metrics_df = metrics_df.with_columns(\n",
    "            pl.col(\"Metadata_Feature_Type\").alias(\"Classifier_type\"),\n",
    "            # Extract the substring that:\n",
    "            #  1. Has a digit (\\d) immediately before it (anchors the match at a number)\n",
    "            #  2. Starts with 'A' and then as few characters as needed (A.*?), captured as group 1\n",
    "            #  3. Stops right before the literal 'T'\n",
    "            pl.col(\"Plate\").str.extract(r\"\\d(A.*?)T\", 1).alias(\"Allele_set\"),\n",
    "            pl.col(\"Full_Classifier_ID\").str.split(\"_\").list.last().alias(\"Batch\")\n",
    "        )\n",
    "    \n",
    "        metrics_df = metrics_df.join(\n",
    "            img_well_qc_sum_dict[bio_rep].select(pl.col(\"plate\", \"well\", \"channel\", \"is_bg\")),\n",
    "            left_on=[\"Plate\", \"well_0\", \"Metadata_Feature_Type\"],\n",
    "            right_on=[\"plate\", \"well\", \"channel\"]\n",
    "        ).rename({\"is_bg\": \"well_0_is_bg\"})\n",
    "        metrics_df = metrics_df.join(\n",
    "            img_well_qc_sum_dict[bio_rep].select(pl.col(\"plate\", \"well\", \"channel\", \"is_bg\")),\n",
    "            left_on=[\"Plate\", \"well_1\", \"Metadata_Feature_Type\"],\n",
    "            right_on=[\"plate\", \"well\", \"channel\"]\n",
    "        ).rename({\"is_bg\": \"well_1_is_bg\"})\n",
    "            \n",
    "        print(\"==========================================================================\")\n",
    "        metric_df_dict[f\"{batch_id}_met\"] = metrics_df\n",
    "        # print(metrics_df_thres.head())\n",
    "        metrics_df_thres = metrics_df\n",
    "        \n",
    "        ## Must be at least min_class_num classifiers per batch\n",
    "        ## Number of classifiers is the same for localization and morph, so just use morph\n",
    "        classifier_count = (\n",
    "            metrics_df_thres.filter(\n",
    "                (~pl.col(\"Metadata_Control\")) & \\\n",
    "                (~pl.col(\"well_0_is_bg\")) & (~pl.col(\"well_1_is_bg\")) & \\\n",
    "                (pl.col(\"Training_imbalance\") < TRN_IMBAL_THRES)\n",
    "            )\n",
    "            .group_by([\"allele_0\", \"Allele_set\", \"Batch\", \"allele_1\"])\n",
    "            .agg([pl.len().alias(\"Number_classifiers\")])\n",
    "        )\n",
    "        classifier_count = classifier_count.pivot(\n",
    "            index=[\"allele_0\", \"allele_1\", \"Allele_set\"],\n",
    "            on=\"Batch\",\n",
    "            values=\"Number_classifiers\",\n",
    "        )\n",
    "        print(f\"Total number of unique classifiers for GFP:\", classifier_count.shape[0])\n",
    "        print(\"Total number of unique variant alleles:\", len(classifier_count.select(\"allele_0\").to_series().unique().to_list()))\n",
    "        print(\"Total number of unique WT genes:\", len(classifier_count.select(\"allele_1\").to_series().unique().to_list()))\n",
    "        print(\"==========================================================================\")\n",
    "    \n",
    "        # Must be at least min_class_num classifiers per batch\n",
    "        # Number of classifiers is the same for localization and morph, so just use morph\n",
    "        classifier_count = (\n",
    "            metrics_df_thres.filter(\n",
    "                (pl.col(\"Training_imbalance\") < TRN_IMBAL_THRES)\n",
    "                & (~pl.col(\"Metadata_Control\"))\n",
    "                & (pl.col(\"Classifier_type\") == \"GFP\")\n",
    "            )\n",
    "            .group_by([\"allele_0\", \"Allele_set\", \"Batch\", \"allele_1\"])\n",
    "            .agg([pl.len().alias(\"Number_classifiers\")])\n",
    "        )\n",
    "        classifier_count = classifier_count.pivot(\n",
    "            index=[\"allele_0\", \"allele_1\", \"Allele_set\"],\n",
    "            on=\"Batch\",\n",
    "            values=\"Number_classifiers\",\n",
    "        )\n",
    "        # display(classifier_count)\n",
    "        # break\n",
    "        print(f\"After filtering out classifiers with training imbalance > {TRN_IMBAL_THRES}:\")\n",
    "        print(\"Total number of unique classifiers:\", classifier_count.shape)\n",
    "        print(\"Total number of unique variant alleles:\", len(classifier_count.select(\"allele_0\").to_series().unique().to_list()))\n",
    "        print(\"Total number of unique WT genes:\", len(classifier_count.select(\"allele_1\").to_series().unique().to_list()))\n",
    "        print(\"==========================================================================\")\n",
    "    \n",
    "        # classifier_count = classifier_count.filter(\n",
    "        #     (pl.col(batch_id) >= MIN_CLASS_NUM)\n",
    "        # )\n",
    "        # print(\"After filtering out alleles with available number of classifiers < 2:\")\n",
    "        # print(\"Total number of unique classifiers:\", classifier_count.shape)\n",
    "        # print(\"Total number of unique variant alleles:\", len(classifier_count.select(\"allele_0\").to_series().unique().to_list()))\n",
    "        # print(\"Total number of unique WT genes:\", len(classifier_count.select(\"allele_1\").to_series().unique().to_list()))\n",
    "    \n",
    "        # # filter based on this\n",
    "        # keep_alleles = classifier_count.select(\"allele_0\").to_series().unique().to_list()\n",
    "        # metrics_df_thres = metrics_df_thres.filter(\n",
    "        #     ~((~pl.col(\"Metadata_Control\")) & ~pl.col(\"allele_0\").is_in(keep_alleles))\n",
    "        # )\n",
    "        # Filter by imbalance and calculate mean AUROC for each batch\n",
    "        metrics_wtvar = (\n",
    "            (\n",
    "                metrics_df_thres.filter(\n",
    "                    (~pl.col(\"well_0_is_bg\")) & (~pl.col(\"well_1_is_bg\")) & \\\n",
    "                    (pl.col(\"Training_imbalance\") < TRN_IMBAL_THRES) & (~pl.col(\"Metadata_Control\"))\n",
    "                )\n",
    "            )\n",
    "            .select([\n",
    "                \"AUROC\",\n",
    "                \"Classifier_type\",\n",
    "                \"Batch\",\n",
    "                \"allele_0\",\n",
    "                \"trainsize_0\",\n",
    "                \"testsize_0\",\n",
    "                \"trainsize_1\",\n",
    "                \"testsize_1\",\n",
    "                \"Allele_set\",\n",
    "                \"Training_imbalance\",\n",
    "            ])#+[f\"AUROC_thresh_{auroc_thres*100:.0f}\" for auroc_thres in AUROC_THRESHOLDS])\n",
    "            .group_by([\"Classifier_type\", \"allele_0\", \"Allele_set\", \"Batch\"]) ## +[f\"AUROC_thresh_{auroc_thres*100:.0f}\" for auroc_thres in AUROC_THRESHOLDS]\n",
    "            .agg([\n",
    "                pl.all()\n",
    "                .exclude([\"Classifier_type\", \"allele_0\", \"Allele_set\", \"Batch\"]) ## +[f\"AUROC_thresh_{auroc_thres*100:.0f}\" for auroc_thres in AUROC_THRESHOLDS]\n",
    "                .mean()\n",
    "                .name.suffix(\"_mean\")\n",
    "            ])\n",
    "        )\n",
    "        metric_df_dict[f\"{batch_id}_met_thres\"] = metrics_df_thres\n",
    "        metric_df_dict[f\"{batch_id}_allele_summary\"] = metrics_wtvar\n",
    "\n",
    "        # os.makedirs(f\"{CLASS_SUMMARY_DIR}/{bio_rep}/{batch_id}\", exist_ok=True)\n",
    "        # metrics_wtvar.write_csv(f\"{CLASS_SUMMARY_DIR}/{bio_rep}/{batch_id}/metrics_summary.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "1d4914be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (1_335, 2)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>allele_0</th><th>AUROC_Mean_GFP_Adj</th></tr><tr><td>str</td><td>f64</td></tr></thead><tbody><tr><td>&quot;GBA_Asp176Asn&quot;</td><td>0.857824</td></tr><tr><td>&quot;KLHL3_Cys164Phe&quot;</td><td>0.860255</td></tr><tr><td>&quot;CCM2_Ala295Thr&quot;</td><td>0.849189</td></tr><tr><td>&quot;CCM2_Glu290Lys&quot;</td><td>0.807653</td></tr><tr><td>&quot;DES_Leu274Pro&quot;</td><td>0.929777</td></tr><tr><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;PSEN1_Arg269His&quot;</td><td>0.997795</td></tr><tr><td>&quot;COMP_Asp319Val&quot;</td><td>0.959706</td></tr><tr><td>&quot;CDKN1A_Asp149Gly&quot;</td><td>0.967215</td></tr><tr><td>&quot;CCM2_Ala205Thr&quot;</td><td>0.802699</td></tr><tr><td>&quot;MAP2K2_Lys61Asn&quot;</td><td>0.848298</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (1_335, 2)\n",
       "┌──────────────────┬────────────────────┐\n",
       "│ allele_0         ┆ AUROC_Mean_GFP_Adj │\n",
       "│ ---              ┆ ---                │\n",
       "│ str              ┆ f64                │\n",
       "╞══════════════════╪════════════════════╡\n",
       "│ GBA_Asp176Asn    ┆ 0.857824           │\n",
       "│ KLHL3_Cys164Phe  ┆ 0.860255           │\n",
       "│ CCM2_Ala295Thr   ┆ 0.849189           │\n",
       "│ CCM2_Glu290Lys   ┆ 0.807653           │\n",
       "│ DES_Leu274Pro    ┆ 0.929777           │\n",
       "│ …                ┆ …                  │\n",
       "│ PSEN1_Arg269His  ┆ 0.997795           │\n",
       "│ COMP_Asp319Val   ┆ 0.959706           │\n",
       "│ CDKN1A_Asp149Gly ┆ 0.967215           │\n",
       "│ CCM2_Ala205Thr   ┆ 0.802699           │\n",
       "│ MAP2K2_Lys61Asn  ┆ 0.848298           │\n",
       "└──────────────────┴────────────────────┘"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele_df = pl.DataFrame()\n",
    "for key, df in metric_df_dict.items():\n",
    "    if \"allele_summary\" in key:\n",
    "        allele_df = pl.concat([allele_df, df], how=\"diagonal_relaxed\")\n",
    "        \n",
    "allele_df = allele_df.filter(\n",
    "    ~((pl.col(\"allele_0\")==\"AGXT_Asp201Asn\")&((pl.col(\"Batch\")==\"B7\")|(pl.col(\"Batch\")==\"B8\"))) &\n",
    "    ~((pl.col(\"allele_0\")==\"MVK_Pro288Leu\")&((pl.col(\"Batch\")==\"B7\")|(pl.col(\"Batch\")==\"B8\"))) &\n",
    "    ~((pl.col(\"allele_0\")==\"GSS_Arg125Cys\")&((pl.col(\"Batch\")==\"B7\")|(pl.col(\"Batch\")==\"B8\"))) &\n",
    "    ~((pl.col(\"allele_0\")==\"GSS_Arg125Cys\")&((pl.col(\"Batch\")==\"B11\")|(pl.col(\"Batch\")==\"B12\"))) &\n",
    "    ~((pl.col(\"allele_0\")==\"HPRT1_His204Asp\")&((pl.col(\"Batch\")==\"B7\")|(pl.col(\"Batch\")==\"B8\"))) &\n",
    "    ~((pl.col(\"allele_0\")==\"MVK_Ser329Asn\")&((pl.col(\"Batch\")==\"B7\")|(pl.col(\"Batch\")==\"B8\"))) &\n",
    "    ~((pl.col(\"allele_0\")==\"RAD51D_Arg165Gln\")&((pl.col(\"Batch\")==\"B7\")|(pl.col(\"Batch\")==\"B8\"))) &\n",
    "    ~((pl.col(\"allele_0\")==\"BCL10_Leu8Leu\")&((pl.col(\"Batch\")==\"B7\")|(pl.col(\"Batch\")==\"B8\")))\n",
    ")\n",
    "allele_df.group_by(\"allele_0\").agg(\n",
    "    pl.col(\"AUROC_mean\").mean().alias(\"AUROC_Mean_GFP_Adj\"),\n",
    ").filter(\n",
    "    # pl.col(\"AUROC_Mean_GFP_Adj\") > 0.9577,\n",
    "    ~pl.col(\"allele_0\").str.contains(\"KCNJ2|STXBP1\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "bd3ac119",
   "metadata": {},
   "outputs": [],
   "source": [
    "allele_df.group_by(\"allele_0\").agg(\n",
    "    pl.col(\"AUROC_mean\").mean().alias(\"AUROC_Mean_GFP_Adj\"),\n",
    ").filter(\n",
    "    # pl.col(\"AUROC_Mean_GFP_Adj\") > 0.9577,\n",
    "    ~pl.col(\"allele_0\").str.contains(\"KCNJ2|STXBP1\")\n",
    ").write_csv(\"../outputs/2.classification_results/imaging_analyses_classification_summary_all_gfp_adj_auroc.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "varchamp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
