{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4ead3db3",
   "metadata": {},
   "source": [
    "# Visualization Variant Painting Images and Cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6e136e31",
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl\n",
    "import os\n",
    "import matplotlib as mpl\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "from skimage.io import imread\n",
    "from tqdm import tqdm\n",
    "import re\n",
    "\n",
    "CP_GALLERY_DIR = \"../../../../2_analysis/3_cytoself/cellpainting_gallery\"\n",
    "batches = [\"2024_01_23_Batch_7\", \"2024_02_06_Batch_8\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e04ba49a",
   "metadata": {},
   "source": [
    "## Loading meta data and variant classification"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d04b9dae",
   "metadata": {},
   "source": [
    "### Read in the meta data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "74170f09",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (5, 30)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>plate_map_name</th><th>well_position</th><th>control_type</th><th>symbol</th><th>gene_allele</th><th>imaging_well</th><th>imaging_plate_R1</th><th>imaging_plate_R2</th><th>node_type</th><th>orf_id_wt</th><th>ccsb_mutation_id</th><th>nt_change_cdna</th><th>aa_change</th><th>allele_set</th><th>imaging_plate</th><th>batch1</th><th>batch2</th><th>clone_ID</th><th>hgmd_ID</th><th>clinvar_allele_ID</th><th>clinvar_clinical_significance</th><th>vector</th><th>notes</th><th>plate</th><th>quad</th><th>c96</th><th>r96</th><th>r384</th><th>c384</th><th>r384_letter</th></tr><tr><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>f64</td><td>f64</td><td>str</td><td>str</td><td>f64</td><td>i64</td><td>i64</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>f64</td><td>str</td><td>i64</td><td>i64</td><td>i64</td><td>i64</td><td>i64</td><td>i64</td><td>str</td></tr></thead><tbody><tr><td>&quot;B7A1R1_P1&quot;</td><td>&quot;A01&quot;</td><td>&quot;disease_wt&quot;</td><td>&quot;ACSF3&quot;</td><td>&quot;ACSF3&quot;</td><td>&quot;A01&quot;</td><td>&quot;B7A1R1_P1&quot;</td><td>&quot;B8A1R2_P1&quot;</td><td>&quot;disease_wt&quot;</td><td>71337.0</td><td>null</td><td>null</td><td>null</td><td>1.0</td><td>1</td><td>7</td><td>8</td><td>null</td><td>null</td><td>null</td><td>null</td><td>527.0</td><td>null</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>1</td><td>&quot;A&quot;</td></tr><tr><td>&quot;B7A1R1_P1&quot;</td><td>&quot;C01&quot;</td><td>&quot;allele&quot;</td><td>&quot;ACSF3&quot;</td><td>&quot;ACSF3_Ala197Thr&quot;</td><td>&quot;C01&quot;</td><td>&quot;B7A1R1_P1&quot;</td><td>&quot;B8A1R2_P1&quot;</td><td>&quot;allele&quot;</td><td>71337.0</td><td>3943.0</td><td>&quot;589G&gt;A&quot;</td><td>&quot;Ala197Thr&quot;</td><td>1.0</td><td>1</td><td>7</td><td>8</td><td>null</td><td>null</td><td>null</td><td>null</td><td>527.0</td><td>null</td><td>1</td><td>1</td><td>1</td><td>2</td><td>3</td><td>1</td><td>&quot;C&quot;</td></tr><tr><td>&quot;B7A1R1_P1&quot;</td><td>&quot;E01&quot;</td><td>&quot;allele&quot;</td><td>&quot;ACSF3&quot;</td><td>&quot;ACSF3_Thr358Ile&quot;</td><td>&quot;E01&quot;</td><td>&quot;B7A1R1_P1&quot;</td><td>&quot;B8A1R2_P1&quot;</td><td>&quot;allele&quot;</td><td>71337.0</td><td>205771.0</td><td>&quot;1073C&gt;T&quot;</td><td>&quot;Thr358Ile&quot;</td><td>1.0</td><td>1</td><td>7</td><td>8</td><td>null</td><td>null</td><td>null</td><td>null</td><td>527.0</td><td>null</td><td>1</td><td>1</td><td>1</td><td>3</td><td>5</td><td>1</td><td>&quot;E&quot;</td></tr><tr><td>&quot;B7A1R1_P1&quot;</td><td>&quot;G01&quot;</td><td>&quot;allele&quot;</td><td>&quot;ACSF3&quot;</td><td>&quot;ACSF3_Met198Arg&quot;</td><td>&quot;G01&quot;</td><td>&quot;B7A1R1_P1&quot;</td><td>&quot;B8A1R2_P1&quot;</td><td>&quot;allele&quot;</td><td>71337.0</td><td>3944.0</td><td>&quot;593T&gt;G&quot;</td><td>&quot;Met198Arg&quot;</td><td>1.0</td><td>1</td><td>7</td><td>8</td><td>null</td><td>null</td><td>null</td><td>null</td><td>527.0</td><td>null</td><td>1</td><td>1</td><td>1</td><td>4</td><td>7</td><td>1</td><td>&quot;G&quot;</td></tr><tr><td>&quot;B7A1R1_P1&quot;</td><td>&quot;I01&quot;</td><td>&quot;allele&quot;</td><td>&quot;ACSF3&quot;</td><td>&quot;ACSF3_Glu359Lys&quot;</td><td>&quot;I01&quot;</td><td>&quot;B7A1R1_P1&quot;</td><td>&quot;B8A1R2_P1&quot;</td><td>&quot;allele&quot;</td><td>71337.0</td><td>205770.0</td><td>&quot;1075G&gt;A&quot;</td><td>&quot;Glu359Lys&quot;</td><td>1.0</td><td>1</td><td>7</td><td>8</td><td>null</td><td>null</td><td>null</td><td>null</td><td>527.0</td><td>null</td><td>1</td><td>1</td><td>1</td><td>5</td><td>9</td><td>1</td><td>&quot;I&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (5, 30)\n",
       "┌────────────────┬───────────────┬──────────────┬────────┬───┬─────┬──────┬──────┬─────────────┐\n",
       "│ plate_map_name ┆ well_position ┆ control_type ┆ symbol ┆ … ┆ r96 ┆ r384 ┆ c384 ┆ r384_letter │\n",
       "│ ---            ┆ ---           ┆ ---          ┆ ---    ┆   ┆ --- ┆ ---  ┆ ---  ┆ ---         │\n",
       "│ str            ┆ str           ┆ str          ┆ str    ┆   ┆ i64 ┆ i64  ┆ i64  ┆ str         │\n",
       "╞════════════════╪═══════════════╪══════════════╪════════╪═══╪═════╪══════╪══════╪═════════════╡\n",
       "│ B7A1R1_P1      ┆ A01           ┆ disease_wt   ┆ ACSF3  ┆ … ┆ 1   ┆ 1    ┆ 1    ┆ A           │\n",
       "│ B7A1R1_P1      ┆ C01           ┆ allele       ┆ ACSF3  ┆ … ┆ 2   ┆ 3    ┆ 1    ┆ C           │\n",
       "│ B7A1R1_P1      ┆ E01           ┆ allele       ┆ ACSF3  ┆ … ┆ 3   ┆ 5    ┆ 1    ┆ E           │\n",
       "│ B7A1R1_P1      ┆ G01           ┆ allele       ┆ ACSF3  ┆ … ┆ 4   ┆ 7    ┆ 1    ┆ G           │\n",
       "│ B7A1R1_P1      ┆ I01           ┆ allele       ┆ ACSF3  ┆ … ┆ 5   ┆ 9    ┆ 1    ┆ I           │\n",
       "└────────────────┴───────────────┴──────────────┴────────┴───┴─────┴──────┴──────┴─────────────┘"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (5, 30)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>plate_map_name</th><th>well_position</th><th>control_type</th><th>symbol</th><th>gene_allele</th><th>imaging_well</th><th>imaging_plate_R1</th><th>imaging_plate_R2</th><th>node_type</th><th>orf_id_wt</th><th>ccsb_mutation_id</th><th>nt_change_cdna</th><th>aa_change</th><th>allele_set</th><th>imaging_plate</th><th>batch1</th><th>batch2</th><th>clone_ID</th><th>hgmd_ID</th><th>clinvar_allele_ID</th><th>clinvar_clinical_significance</th><th>vector</th><th>notes</th><th>plate</th><th>quad</th><th>c96</th><th>r96</th><th>r384</th><th>c384</th><th>r384_letter</th></tr><tr><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>f64</td><td>f64</td><td>str</td><td>str</td><td>f64</td><td>i64</td><td>i64</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>f64</td><td>str</td><td>i64</td><td>i64</td><td>i64</td><td>i64</td><td>i64</td><td>i64</td><td>str</td></tr></thead><tbody><tr><td>&quot;B7A1R1_P2&quot;</td><td>&quot;H24&quot;</td><td>&quot;disease_wt&quot;</td><td>&quot;MAP2K2&quot;</td><td>&quot;MAP2K2&quot;</td><td>&quot;H24&quot;</td><td>&quot;B7A1R1_P2&quot;</td><td>&quot;B8A1R2_P2&quot;</td><td>&quot;disease_wt&quot;</td><td>6666.0</td><td>null</td><td>null</td><td>null</td><td>1.0</td><td>2</td><td>7</td><td>8</td><td>null</td><td>null</td><td>null</td><td>null</td><td>527.0</td><td>null</td><td>2</td><td>4</td><td>12</td><td>4</td><td>8</td><td>24</td><td>&quot;H&quot;</td></tr><tr><td>&quot;B7A1R1_P2&quot;</td><td>&quot;J24&quot;</td><td>&quot;allele&quot;</td><td>&quot;MAP2K2&quot;</td><td>&quot;MAP2K2_Thr17Ala&quot;</td><td>&quot;J24&quot;</td><td>&quot;B7A1R1_P2&quot;</td><td>&quot;B8A1R2_P2&quot;</td><td>&quot;allele&quot;</td><td>6666.0</td><td>51.0</td><td>&quot;49A&gt;G&quot;</td><td>&quot;Thr17Ala&quot;</td><td>1.0</td><td>2</td><td>7</td><td>8</td><td>null</td><td>null</td><td>null</td><td>null</td><td>527.0</td><td>null</td><td>2</td><td>4</td><td>12</td><td>5</td><td>10</td><td>24</td><td>&quot;J&quot;</td></tr><tr><td>&quot;B7A1R1_P2&quot;</td><td>&quot;L24&quot;</td><td>&quot;disease_wt&quot;</td><td>&quot;MAX&quot;</td><td>&quot;MAX&quot;</td><td>&quot;L24&quot;</td><td>&quot;B7A1R1_P2&quot;</td><td>&quot;B8A1R2_P2&quot;</td><td>&quot;disease_wt&quot;</td><td>4384.0</td><td>null</td><td>null</td><td>null</td><td>1.0</td><td>2</td><td>7</td><td>8</td><td>null</td><td>null</td><td>null</td><td>null</td><td>527.0</td><td>null</td><td>2</td><td>4</td><td>12</td><td>6</td><td>12</td><td>24</td><td>&quot;L&quot;</td></tr><tr><td>&quot;B7A1R1_P2&quot;</td><td>&quot;N24&quot;</td><td>&quot;disease_wt&quot;</td><td>&quot;MCEE&quot;</td><td>&quot;MCEE&quot;</td><td>&quot;N24&quot;</td><td>&quot;B7A1R1_P2&quot;</td><td>&quot;B8A1R2_P2&quot;</td><td>&quot;disease_wt&quot;</td><td>13113.0</td><td>null</td><td>null</td><td>null</td><td>1.0</td><td>2</td><td>7</td><td>8</td><td>null</td><td>null</td><td>null</td><td>null</td><td>527.0</td><td>null</td><td>2</td><td>4</td><td>12</td><td>7</td><td>14</td><td>24</td><td>&quot;N&quot;</td></tr><tr><td>&quot;B7A1R1_P2&quot;</td><td>&quot;P24&quot;</td><td>&quot;allele&quot;</td><td>&quot;NAA10&quot;</td><td>&quot;NAA10_Arg195Cys&quot;</td><td>&quot;P24&quot;</td><td>&quot;B7A1R1_P2&quot;</td><td>&quot;B8A1R2_P2&quot;</td><td>&quot;allele&quot;</td><td>3180.0</td><td>243.0</td><td>&quot;583C&gt;T&quot;</td><td>&quot;Arg195Cys&quot;</td><td>1.0</td><td>2</td><td>7</td><td>8</td><td>null</td><td>null</td><td>null</td><td>null</td><td>527.0</td><td>null</td><td>2</td><td>4</td><td>12</td><td>8</td><td>16</td><td>24</td><td>&quot;P&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (5, 30)\n",
       "┌────────────────┬───────────────┬──────────────┬────────┬───┬─────┬──────┬──────┬─────────────┐\n",
       "│ plate_map_name ┆ well_position ┆ control_type ┆ symbol ┆ … ┆ r96 ┆ r384 ┆ c384 ┆ r384_letter │\n",
       "│ ---            ┆ ---           ┆ ---          ┆ ---    ┆   ┆ --- ┆ ---  ┆ ---  ┆ ---         │\n",
       "│ str            ┆ str           ┆ str          ┆ str    ┆   ┆ i64 ┆ i64  ┆ i64  ┆ str         │\n",
       "╞════════════════╪═══════════════╪══════════════╪════════╪═══╪═════╪══════╪══════╪═════════════╡\n",
       "│ B7A1R1_P2      ┆ H24           ┆ disease_wt   ┆ MAP2K2 ┆ … ┆ 4   ┆ 8    ┆ 24   ┆ H           │\n",
       "│ B7A1R1_P2      ┆ J24           ┆ allele       ┆ MAP2K2 ┆ … ┆ 5   ┆ 10   ┆ 24   ┆ J           │\n",
       "│ B7A1R1_P2      ┆ L24           ┆ disease_wt   ┆ MAX    ┆ … ┆ 6   ┆ 12   ┆ 24   ┆ L           │\n",
       "│ B7A1R1_P2      ┆ N24           ┆ disease_wt   ┆ MCEE   ┆ … ┆ 7   ┆ 14   ┆ 24   ┆ N           │\n",
       "│ B7A1R1_P2      ┆ P24           ┆ allele       ┆ NAA10  ┆ … ┆ 8   ┆ 16   ┆ 24   ┆ P           │\n",
       "└────────────────┴───────────────┴──────────────┴────────┴───┴─────┴──────┴──────┴─────────────┘"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "meta_data = pl.DataFrame()\n",
    "plate_map_dirs = [\n",
    "    f\"../snakemake_pipeline/inputs/metadata/platemaps/{batch_id}/platemap/\" \n",
    "    for batch_id in batches\n",
    "]\n",
    "for plate_map_dir in plate_map_dirs:\n",
    "    for plate_map in os.listdir(plate_map_dir):\n",
    "        if plate_map.endswith(\".txt\") and plate_map.startswith(\"B7A1R1_P\"):\n",
    "            plate_map_path = os.path.join(plate_map_dir, plate_map)\n",
    "            plate_map_data = pl.read_csv(plate_map_path, separator=\"\\t\", infer_schema_length=1000000)\n",
    "            # print(plate_map_data)\n",
    "            if meta_data.is_empty():\n",
    "                meta_data = plate_map_data\n",
    "            else:\n",
    "                meta_data = pl.concat([meta_data, plate_map_data], how=\"vertical_relaxed\")\n",
    "\n",
    "display(meta_data.head())\n",
    "display(meta_data.tail())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a87e311f",
   "metadata": {},
   "source": [
    "### Classification of variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ee28d2cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (5, 12)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>Classifier_type</th><th>allele_0</th><th>Allele_set</th><th>Batch</th><th>AUROC_thresh</th><th>AUROC_mean</th><th>trainsize_0_mean</th><th>testsize_0_mean</th><th>trainsize_1_mean</th><th>testsize_1_mean</th><th>Training_imbalance_mean</th><th>morph_change</th></tr><tr><td>str</td><td>str</td><td>str</td><td>str</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>i32</td></tr></thead><tbody><tr><td>&quot;morphology&quot;</td><td>&quot;HBD_Gly26Asp&quot;</td><td>&quot;A1P2&quot;</td><td>&quot;B7&quot;</td><td>0.997315</td><td>0.994581</td><td>1640.666667</td><td>648.333333</td><td>621.666667</td><td>202.333333</td><td>2.644345</td><td>0</td></tr><tr><td>&quot;morphology&quot;</td><td>&quot;PYGL_Asn339Ser&quot;</td><td>&quot;A1P3&quot;</td><td>&quot;B8&quot;</td><td>0.99807</td><td>0.459475</td><td>230.666667</td><td>115.333333</td><td>185.333333</td><td>92.666667</td><td>1.240988</td><td>0</td></tr><tr><td>&quot;morphology&quot;</td><td>&quot;AGXT_Ala186Val&quot;</td><td>&quot;A1P1&quot;</td><td>&quot;B7&quot;</td><td>0.997315</td><td>0.852631</td><td>1099.5</td><td>366.5</td><td>1590.75</td><td>530.25</td><td>1.448952</td><td>0</td></tr><tr><td>&quot;morphology&quot;</td><td>&quot;PKP2_Tyr582Asn&quot;</td><td>&quot;A1P3&quot;</td><td>&quot;B8&quot;</td><td>0.99807</td><td>0.466434</td><td>111.333333</td><td>55.666667</td><td>78.666667</td><td>39.333333</td><td>1.635455</td><td>0</td></tr><tr><td>&quot;morphology&quot;</td><td>&quot;PKP2_Thr482Met&quot;</td><td>&quot;A1P3&quot;</td><td>&quot;B7&quot;</td><td>0.997315</td><td>0.689049</td><td>645.666667</td><td>262.333333</td><td>395.333333</td><td>71.666667</td><td>1.635137</td><td>0</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (5, 12)\n",
       "┌────────────┬────────────┬────────────┬───────┬───┬───────────┬───────────┬───────────┬───────────┐\n",
       "│ Classifier ┆ allele_0   ┆ Allele_set ┆ Batch ┆ … ┆ trainsize ┆ testsize_ ┆ Training_ ┆ morph_cha │\n",
       "│ _type      ┆ ---        ┆ ---        ┆ ---   ┆   ┆ _1_mean   ┆ 1_mean    ┆ imbalance ┆ nge       │\n",
       "│ ---        ┆ str        ┆ str        ┆ str   ┆   ┆ ---       ┆ ---       ┆ _mean     ┆ ---       │\n",
       "│ str        ┆            ┆            ┆       ┆   ┆ f64       ┆ f64       ┆ ---       ┆ i32       │\n",
       "│            ┆            ┆            ┆       ┆   ┆           ┆           ┆ f64       ┆           │\n",
       "╞════════════╪════════════╪════════════╪═══════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡\n",
       "│ morphology ┆ HBD_Gly26A ┆ A1P2       ┆ B7    ┆ … ┆ 621.66666 ┆ 202.33333 ┆ 2.644345  ┆ 0         │\n",
       "│            ┆ sp         ┆            ┆       ┆   ┆ 7         ┆ 3         ┆           ┆           │\n",
       "│ morphology ┆ PYGL_Asn33 ┆ A1P3       ┆ B8    ┆ … ┆ 185.33333 ┆ 92.666667 ┆ 1.240988  ┆ 0         │\n",
       "│            ┆ 9Ser       ┆            ┆       ┆   ┆ 3         ┆           ┆           ┆           │\n",
       "│ morphology ┆ AGXT_Ala18 ┆ A1P1       ┆ B7    ┆ … ┆ 1590.75   ┆ 530.25    ┆ 1.448952  ┆ 0         │\n",
       "│            ┆ 6Val       ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ morphology ┆ PKP2_Tyr58 ┆ A1P3       ┆ B8    ┆ … ┆ 78.666667 ┆ 39.333333 ┆ 1.635455  ┆ 0         │\n",
       "│            ┆ 2Asn       ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ morphology ┆ PKP2_Thr48 ┆ A1P3       ┆ B7    ┆ … ┆ 395.33333 ┆ 71.666667 ┆ 1.635137  ┆ 0         │\n",
       "│            ┆ 2Met       ┆            ┆       ┆   ┆ 3         ┆           ┆           ┆           │\n",
       "└────────────┴────────────┴────────────┴───────┴───┴───────────┴───────────┴───────────┴───────────┘"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (5, 12)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>Classifier_type</th><th>allele_0</th><th>Allele_set</th><th>Batch</th><th>AUROC_thresh</th><th>AUROC_mean</th><th>trainsize_0_mean</th><th>testsize_0_mean</th><th>trainsize_1_mean</th><th>testsize_1_mean</th><th>Training_imbalance_mean</th><th>mislocalized</th></tr><tr><td>str</td><td>str</td><td>str</td><td>str</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>i32</td></tr></thead><tbody><tr><td>&quot;localization&quot;</td><td>&quot;AGXT_Ala210Pro&quot;</td><td>&quot;A1P1&quot;</td><td>&quot;B7&quot;</td><td>0.71418</td><td>0.932789</td><td>1238.25</td><td>412.75</td><td>1590.75</td><td>530.25</td><td>1.283454</td><td>1</td></tr><tr><td>&quot;localization&quot;</td><td>&quot;AGXT_Gly161Arg&quot;</td><td>&quot;A1P1&quot;</td><td>&quot;B8&quot;</td><td>0.861392</td><td>0.80332</td><td>265.5</td><td>88.5</td><td>440.25</td><td>146.75</td><td>1.734075</td><td>0</td></tr><tr><td>&quot;localization&quot;</td><td>&quot;COMP_Arg718Pro&quot;</td><td>&quot;A1P1&quot;</td><td>&quot;B8&quot;</td><td>0.861392</td><td>0.968708</td><td>369.75</td><td>123.25</td><td>201.0</td><td>67.0</td><td>1.847131</td><td>1</td></tr><tr><td>&quot;localization&quot;</td><td>&quot;KRT86_Ala118Glu&quot;</td><td>&quot;A1P2&quot;</td><td>&quot;B8&quot;</td><td>0.861392</td><td>0.692461</td><td>208.5</td><td>69.5</td><td>329.25</td><td>109.75</td><td>1.598304</td><td>0</td></tr><tr><td>&quot;localization&quot;</td><td>&quot;TH_Ile209Met&quot;</td><td>&quot;A1P4&quot;</td><td>&quot;B8&quot;</td><td>0.861392</td><td>0.629974</td><td>259.5</td><td>86.5</td><td>193.5</td><td>64.5</td><td>1.336459</td><td>0</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (5, 12)\n",
       "┌────────────┬────────────┬────────────┬───────┬───┬───────────┬───────────┬───────────┬───────────┐\n",
       "│ Classifier ┆ allele_0   ┆ Allele_set ┆ Batch ┆ … ┆ trainsize ┆ testsize_ ┆ Training_ ┆ mislocali │\n",
       "│ _type      ┆ ---        ┆ ---        ┆ ---   ┆   ┆ _1_mean   ┆ 1_mean    ┆ imbalance ┆ zed       │\n",
       "│ ---        ┆ str        ┆ str        ┆ str   ┆   ┆ ---       ┆ ---       ┆ _mean     ┆ ---       │\n",
       "│ str        ┆            ┆            ┆       ┆   ┆ f64       ┆ f64       ┆ ---       ┆ i32       │\n",
       "│            ┆            ┆            ┆       ┆   ┆           ┆           ┆ f64       ┆           │\n",
       "╞════════════╪════════════╪════════════╪═══════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡\n",
       "│ localizati ┆ AGXT_Ala21 ┆ A1P1       ┆ B7    ┆ … ┆ 1590.75   ┆ 530.25    ┆ 1.283454  ┆ 1         │\n",
       "│ on         ┆ 0Pro       ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ localizati ┆ AGXT_Gly16 ┆ A1P1       ┆ B8    ┆ … ┆ 440.25    ┆ 146.75    ┆ 1.734075  ┆ 0         │\n",
       "│ on         ┆ 1Arg       ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ localizati ┆ COMP_Arg71 ┆ A1P1       ┆ B8    ┆ … ┆ 201.0     ┆ 67.0      ┆ 1.847131  ┆ 1         │\n",
       "│ on         ┆ 8Pro       ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ localizati ┆ KRT86_Ala1 ┆ A1P2       ┆ B8    ┆ … ┆ 329.25    ┆ 109.75    ┆ 1.598304  ┆ 0         │\n",
       "│ on         ┆ 18Glu      ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "│ localizati ┆ TH_Ile209M ┆ A1P4       ┆ B8    ┆ … ┆ 193.5     ┆ 64.5      ┆ 1.336459  ┆ 0         │\n",
       "│ on         ┆ et         ┆            ┆       ┆   ┆           ┆           ┆           ┆           │\n",
       "└────────────┴────────────┴────────────┴───────┴───┴───────────┴───────────┴───────────┴───────────┘"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# trn_imbalance_thresh = 3\n",
    "\n",
    "# Paths\n",
    "metrics_dir = \"../snakemake_pipeline/outputs/analyses/{}/profiles_tcdropped_filtered_var_mad_outlier_featselect\"\n",
    "metrics_df, metrics_wtvar = pl.DataFrame(), pl.DataFrame()\n",
    "for batch in batches:\n",
    "    met_dir = metrics_dir.format(batch)\n",
    "    metrics_df_batch = pl.read_csv(f\"{met_dir}/metrics.csv\")\n",
    "    metrics_wtvar_batch = pl.read_csv(f\"{met_dir}/metrics_summary.csv\")\n",
    "\n",
    "    metrics_df = pl.concat([metrics_df, metrics_df_batch])\n",
    "    metrics_wtvar = pl.concat([metrics_wtvar, metrics_wtvar_batch])\n",
    "\n",
    "## get the localization metrics\n",
    "local_wtvar = metrics_wtvar.filter(pl.col(\"Classifier_type\") == \"localization\")\n",
    "## get the alleles and their reference genes\n",
    "local_wtvar = local_wtvar.with_columns(\n",
    "    pl.when(pl.col(\"AUROC_mean\") > pl.col(\"AUROC_thresh\"))\n",
    "    .then(1)\n",
    "    .otherwise(0)\n",
    "    .alias(\"mislocalized\")\n",
    ").unique()\n",
    "\n",
    "## get the localization metrics\n",
    "morph_wtvar = metrics_wtvar.filter(pl.col(\"Classifier_type\") == \"morphology\")\n",
    "## get the alleles and their reference genes\n",
    "morph_wtvar = morph_wtvar.with_columns(\n",
    "    pl.when(pl.col(\"AUROC_mean\") > pl.col(\"AUROC_thresh\"))\n",
    "    .then(1)\n",
    "    .otherwise(0)\n",
    "    .alias(\"morph_change\")\n",
    ").unique()\n",
    "\n",
    "display(morph_wtvar.head())\n",
    "display(local_wtvar.head())\n",
    "\n",
    "# misloc_alleles = (\n",
    "#     local_wtvar\n",
    "#     .select(\"allele_0\")\n",
    "#     .to_series()\n",
    "#     .to_list()\n",
    "# )\n",
    "# misloc_binary = (\n",
    "#     local_wtvar.pivot(\n",
    "#         index=[\"allele_0\", \"Allele_set\"], ## , \"AUROC_mean\"\n",
    "#         columns=\"Batch\",\n",
    "#         values=\"AUROC_mean\",\n",
    "#     )\n",
    "#     .rename({\"B7\": \"AUROC_mean_B7\", \"B8\": \"AUROC_mean_B8\"})\n",
    "# )\n",
    "# misloc_binary\n",
    "\n",
    "# misloc_alleles = (\n",
    "#     morph_wtvar\n",
    "#     .select(\"allele_0\")\n",
    "#     .to_series()\n",
    "#     .to_list()\n",
    "# )\n",
    "# morph_binary = (\n",
    "#     morph_wtvar.pivot(\n",
    "#         index=[\"allele_0\", \"Allele_set\"], ## , \"AUROC_mean\"\n",
    "#         columns=\"Batch\",\n",
    "#         values=[\"morph_change\"],\n",
    "#     )\n",
    "#     .with_columns(\n",
    "#         ((pl.col(\"B7\") == 1) & (pl.col(\"B8\") == 1)).alias(\n",
    "#             \"morph_change_both_batches\"\n",
    "#         )\n",
    "#     )\n",
    "#     .rename({\"B7\": \"morph_change_batch7\", \"B8\": \"morph_change_batch8\"})\n",
    "#     .filter(pl.col(\"allele_0\").is_in(ind_class[\"allele_0\"]))\n",
    "# )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "436b3c2b",
   "metadata": {},
   "source": [
    "## Plot Variant Painting Images"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d87f30c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define mapping between simple names and folder names\n",
    "batch_dict = {\n",
    "    \"B7A1R1\": \"2024_01_23_Batch_7\",\n",
    "    \"B7A2R1\": \"2024_01_23_Batch_7\",\n",
    "    \"B8A1R2\": \"2024_02_06_Batch_8\",\n",
    "    \"B8A2R2\": \"2024_02_06_Batch_8\"\n",
    "}\n",
    "\n",
    "letter_dict = {\n",
    "    \"A\": \"01\",\n",
    "    \"B\": \"02\",\n",
    "    \"C\": \"03\",\n",
    "    \"D\": \"04\",\n",
    "    \"E\": \"05\",\n",
    "    \"F\": \"06\",\n",
    "    \"G\": \"07\",\n",
    "    \"H\": \"08\",\n",
    "    \"I\": \"09\",\n",
    "    \"J\": \"10\",\n",
    "    \"K\": \"11\",\n",
    "    \"L\": \"12\",\n",
    "    \"M\": \"13\",\n",
    "    \"N\": \"14\",\n",
    "    \"O\": \"15\",\n",
    "    \"P\": \"16\",\n",
    "}\n",
    "\n",
    "plate_dict = {    \n",
    "    \"B7A1R1_P1\": {\"T1\": \"2024_01_17_B7A1R1_P1T1__2024_01_17T08_35_58_Measurement_1\",\n",
    "                  \"T2\": \"2024_01_17_B7A1R1_P1T2__2024_01_17T10_13_45_Measurement_1\",\n",
    "                  \"T3\": \"2024_01_17_B7A1R1_P1T3__2024_01_17T11_58_08_Measurement_1\",\n",
    "                  \"T4\": \"2024_01_17_B7A1R1_P1T4__2024_01_17T13_45_14_Measurement_1\"},\n",
    "    \n",
    "    \"B7A1R1_P2\": {\"T1\": \"2024_01_17_B7A1R1_P2T1__2024_01_17T15_33_09_Measurement_1\",\n",
    "                  \"T2\": \"2024_01_17_B7A1R1_P2T2__2024_01_18T08_25_01_Measurement_1\",\n",
    "                  \"T3\": \"2024_01_17_B7A1R1_P2T3__2024_01_18T10_47_36_Measurement_1\",\n",
    "                  \"T4\": \"2024_01_17_B7A1R1_P2T4__2024_01_18T12_48_20_Measurement_1\"},\n",
    "        \n",
    "    \"B7A1R1_P3\": {\"T1\": \"2024_01_18_B7A1R1_P3T1__2024_01_18T14_27_08_Measurement_1\",\n",
    "                  \"T2\": \"2024_01_19_B7A1R1_P3T2__2024_01_19T08_23_30_Measurement_1\",\n",
    "                  \"T3\": \"2024_01_19_B7A1R1_P3T3__2024_01_19T10_01_45_Measurement_1\",\n",
    "                  \"T4\": \"2024_01_19_B7A1R1_P3T4__2024_01_19T12_00_10_Measurement_1\"},\n",
    "            \n",
    "    \"B7A1R1_P4\": {\"T1\": \"2024_01_19_B7A1R1_P4T1__2024_01_19T13_50_55_Measurement_1\",\n",
    "                  \"T2\": \"2024_01_23_B7A1R1_P4T2__2024_01_23T10_13_00_Measurement_1\",\n",
    "                  \"T3\": \"2024_01_22_B7A1R1_P4T3__2024_01_22T08_37_41_Measurement_1\",\n",
    "                  \"T4\": \"2024_01_22_B7A1R1_P4T4__2024_01_22T10_27_16_Measurement_1\"},\n",
    "    \n",
    "    \"B7A2R1_P1\": {\"T1\": \"2024_01_22_B7A2R1_P1T1__2024_01_22T12_13_13_Measurement_1\",\n",
    "                  \"T2\": \"2024_01_22_B7A2R1_P1T2__2024_01_22T13_52_24_Measurement_1\",\n",
    "                  \"T3\": \"2024_01_22_B7A2R1_P1T3__2024_01_22T15_29_31_Measurement_1\",\n",
    "                  \"T4\": \"2024_01_23_B7A2R1_P1T4__2024_01_23T08_28_07_Measurement_1\"},\n",
    "    \n",
    "    \"B8A1R2_P1\": {\"T1\": \"2024_01_31_B8A1R2_P1T1__2024_01_31T10_11_57_Measurement_1\",\n",
    "                  \"T2\": \"2024_01_31_B8A1R2_P1T2__2024_01_31T08_35_51_Measurement_2\",\n",
    "                  \"T3\": \"2024_01_31_B8A1R2_P1T3__2024_01_31T12_09_14_Measurement_1\",\n",
    "                  \"T4\": \"2024_01_31_B8A1R2_P1T4__2024_01_31T14_02_18_Measurement_2\"},\n",
    "    \n",
    "    \"B8A1R2_P2\": {\"T1\": \"2024_01_31_B8A1R2_P2T1__2024_01_31T15_41_23_Measurement_1\",\n",
    "                  \"T2\": \"2024_02_01_B8A1R2_P2T2__2024_02_01T10_23_20_Measurement_2\",\n",
    "                  \"T3\": \"2024_02_01_B8A1R2_P2T3__2024_02_01T12_16_30_Measurement_4\",\n",
    "                  \"T4\": \"2024_02_01_B8A1R2_P2T4__2024_02_01T14_05_52_Measurement_1\"},\n",
    "        \n",
    "    \"B8A1R2_P3\": {\"T1\": \"2024_02_02_B8A1R2_P3T1__2024_02_02T08_32_30_Measurement_2\",\n",
    "                  \"T2\": \"2024_02_02_B8A1R2_P3T2__2024_02_02T10_08_05_Measurement_1\",\n",
    "                  \"T3\": \"2024_02_02_B8A1R2_P3T3__2024_02_02T11_58_46_Measurement_2\",\n",
    "                  \"T4\": \"2024_02_02_B8A1R2_P3T4__2024_02_02T13_51_50_Measurement_1\"},\n",
    "            \n",
    "    \"B8A1R2_P4\": {\"T1\": \"2024_02_02_B8A1R2_P4T1__2024_02_02T15_32_28_Measurement_1\",\n",
    "                  \"T2\": \"2024_02_05_B8A1R2_P4T2__2024_02_05T08_22_47_Measurement_2\",\n",
    "                  \"T3\": \"2024_02_05_B8A1R2_P4T3__2024_02_05T10_00_30_Measurement_1\",\n",
    "                  \"T4\": \"2024_02_05_B8A1R2_P4T4__2024_02_05T11_38_50_Measurement_1\"},\n",
    "    \n",
    "    \"B8A2R2_P1\": {\"T1\": \"2024_02_05_B8A2R2_P1T1__2024_02_05T13_23_28_Measurement_1\",\n",
    "                  \"T2\": \"2024_02_05_B8A2R2_P1T2__2024_02_05T15_18_32_Measurement_1\",\n",
    "                  \"T3\": \"2024_02_06_B8A2R2_P1T3__2024_02_06T08_23_49_Measurement_1\",\n",
    "                  \"T4\": \"2024_02_06_B8A2R2_P1T4__2024_02_06T10_18_16_Measurement_1\"}\n",
    "}\n",
    "\n",
    "channel_dict = {\n",
    "    \"DAPI\": \"1\",\n",
    "    \"GFP\": \"2\",\n",
    "    \"AGP\": \"3\",\n",
    "    \"Mito\": \"4\",\n",
    "    \"Brightfield1\": \"5\",\n",
    "    \"Brightfield2\": \"6\",\n",
    "    \"Brightfield\": \"7\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1651eb19",
   "metadata": {},
   "outputs": [],
   "source": [
    "tiff_imgs_dir = f\"{CP_GALLERY_DIR}/tiff_images/cpg0020-varchamp/broad/images\"\n",
    "out_imgs_dir = \"../snakemake_pipeline/outputs/visualize_imgs\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a2b8cd12",
   "metadata": {},
   "outputs": [],
   "source": [
    "def channel_to_cmap(channel):\n",
    "    if channel == \"GFP\":\n",
    "        cmap = mpl.colors.LinearSegmentedColormap.from_list(\"green_cmap\", [\"#000\",\"#65fe08\"])\n",
    "    elif channel == \"DAPI\":\n",
    "        cmap = mpl.colors.LinearSegmentedColormap.from_list(\"green_cmap\", [\"#000\",\"#0000FF\"])\n",
    "    elif channel == \"Mito\":\n",
    "        cmap = mpl.colors.LinearSegmentedColormap.from_list(\"green_cmap\", [\"#000\",\"#FF0000\"]) \n",
    "    elif channel == \"AGP\":\n",
    "        cmap = mpl.colors.LinearSegmentedColormap.from_list(\"green_cmap\", [\"#000\",\"#FFFF00\"]) \n",
    "    else:\n",
    "        cmap = \"gray\"\n",
    "    return cmap\n",
    "\n",
    "\n",
    "# Convert letter rows to numbers\n",
    "def well_to_coordinates(well):\n",
    "    row_letter, col_number = re.match(r\"([A-P])(\\d{2})\", well).groups()\n",
    "    row_index = ord(row_letter) - ord('A') + 1  # Convert 'A'->1, 'B'->2, ..., 'P'->16\n",
    "    col_index = int(col_number)  # Convert string column to integer\n",
    "    return well, row_index, col_index\n",
    "\n",
    "\n",
    "# Compute distances from edges and find the most centered well\n",
    "def compute_distance(row, col):\n",
    "    return min(row - 1, 16 - row, col - 1, 24 - col)  # Distance from nearest edge\n",
    "    \n",
    "\n",
    "def plot_allele(pm, variant, sel_channel, auroc_df, site=\"05\", max_intensity=0.99, display=False, imgs_dir=tiff_imgs_dir, output_dir=out_imgs_dir):\n",
    "    cmap = channel_to_cmap(sel_channel)\n",
    "    channel = channel_dict[sel_channel]\n",
    "    auroc = auroc_df.filter(pl.col(\"allele_0\")==variant)[\"AUROC_mean\"].mean()\n",
    "    if os.path.exists(os.path.join(output_dir, f\"{variant}_{sel_channel}_{auroc:.3f}.png\")):\n",
    "        print(f\"Image for {variant} already exists.\")\n",
    "        return None\n",
    "        \n",
    "    ## get the number of wells/images per allele\n",
    "    wt = variant.split(\"_\")[0]\n",
    "    wt_wells = pm.filter(pl.col(\"gene_allele\") == wt).select(\"imaging_well\").to_pandas().values.flatten()\n",
    "    var_wells = pm.filter(pl.col(\"gene_allele\") == variant).select(\"imaging_well\").to_pandas().values.flatten()\n",
    "    plate_map = pm.filter(pl.col(\"gene_allele\") == variant).select(\"plate_map_name\").to_pandas().values.flatten()\n",
    "\n",
    "    if len(wt_wells) > 1:\n",
    "        # Get coordinates of wells\n",
    "        well_coords = [well_to_coordinates(w) for w in set([ref_well_pl for ref_well_pl in wt_wells])]\n",
    "        # Sort wells by max distance from edges (descending)\n",
    "        wt_wells = [max(well_coords, key=lambda x: compute_distance(x[1], x[2]))[0]]\n",
    "\n",
    "    pm_var = pm.filter((pl.col(\"imaging_well\").is_in(np.concatenate([wt_wells, var_wells])))&(pl.col(\"plate_map_name\").is_in(plate_map))).sort(\"control_type\")\n",
    "    # print(pm_var)\n",
    "    plt.clf()\n",
    "\n",
    "    fig, axes = plt.subplots((len(wt_wells)+len(var_wells))*2, 4, figsize=(15, 16), sharex=True, sharey=True)\n",
    "    for wt_var, pm_row in enumerate(pm_var.iter_rows(named=True)):\n",
    "        # print(pm_row)\n",
    "        if pm_row[\"control_type\"] == \"allele\":\n",
    "            well = var_wells[0]\n",
    "            allele = variant\n",
    "        else:\n",
    "            well = wt_wells[0]\n",
    "            allele = wt\n",
    "\n",
    "        for i in range(8):\n",
    "            if i < 4:\n",
    "                sel_plate = pm_row[\"imaging_plate_R1\"]\n",
    "            else:\n",
    "                sel_plate = pm_row[\"imaging_plate_R2\"]\n",
    "                \n",
    "            batch = batch_dict[sel_plate.split(\"_\")[0]]\n",
    "            batch_img_dir = f'{imgs_dir}/{batch}/images'\n",
    "            \n",
    "            letter = well[0]\n",
    "            row = letter_dict[letter]\n",
    "            col = well[1:3]\n",
    "            \n",
    "            plate_img_dir = plate_dict[sel_plate][f\"T{i%4+1}\"]\n",
    "            img_file = f\"r{row}c{col}f{site}p01-ch{channel}sk1fk1fl1.tiff\"\n",
    "            img = imread(f\"{batch_img_dir}/{plate_img_dir}/Images/{img_file}\", as_gray=True)\n",
    "            \n",
    "            # Get key intensity percentiles\n",
    "            p99 = np.percentile(img, 99)\n",
    "            p100 = np.max(img)\n",
    "\n",
    "            # If there's a big jump between 99th and max, lower the threshold\n",
    "            if p100 > 1.5 * p99:\n",
    "                # print(\"More contrast\")\n",
    "                max_intensity = 0.95  # More contrast\n",
    "            else:\n",
    "                max_intensity = max_intensity  # Allow more of the bright tail\n",
    "\n",
    "            plot_idx = i+wt_var*4*2\n",
    "            axes.flatten()[plot_idx].imshow(img, vmin=0, vmax=np.percentile(img, max_intensity*100), cmap=cmap)\n",
    "            plot_label = f\"{sel_channel}:{sel_plate},T{i%4+1}\\nWell:{well},Site:{site}\\n{allele}\"\n",
    "            axes.flatten()[plot_idx].text(0.03, 0.97, plot_label, color='white', fontsize=10,\n",
    "                    verticalalignment='top', horizontalalignment='left', transform=axes.flatten()[plot_idx].transAxes,\n",
    "                    bbox=dict(facecolor='black', alpha=0.3, linewidth=2))\n",
    "            int_95 = str(int(round(np.percentile(img, 95))))\n",
    "            axes.flatten()[plot_idx].text(0.95, 0.05, f\"95th Percentile\\nIntensity:{int_95}\", color='white', fontsize=12,\n",
    "                           verticalalignment='bottom', horizontalalignment='right', transform=axes.flatten()[plot_idx].transAxes)\n",
    "            axes.flatten()[plot_idx].axis(\"off\")\n",
    "        \n",
    "    plt.tight_layout()\n",
    "    plt.subplots_adjust(wspace=.01, hspace=-0.2, top=.99)\n",
    "    if display:\n",
    "        plt.show()\n",
    "    if output_dir:\n",
    "        plt.savefig(os.path.join(output_dir, f\"{variant}_{sel_channel}_{auroc:.3f}.png\"), dpi=400, bbox_inches='tight')\n",
    "        plt.close(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e4f9f766",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot_allele(meta_data, \"SMAD1_Val3Ala\", \"GFP\", auroc_df=local_wtvar, site=\"05\", max_intensity=0.999, display=True, imgs_dir=tiff_imgs_dir, output_dir=out_imgs_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c8eb18f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# variant_list = ['ACSF3_Met266Val',\n",
    "#  'ACSF3_Gly119Asp',\n",
    "#  'ACSF3_Arg558Trp',\n",
    "#  'ACSF3_Arg471Trp',\n",
    "#  'ACSF3_Pro243Leu',\n",
    "#  'ACSF3_Asp457Asn',\n",
    "#  'ACSF3_Thr358Ile']\n",
    "\n",
    "# variant_list = ['PLA2G6_Thr585Ile',\n",
    "#             'PLA2G6_Leu481Gln',\n",
    "#             'PLA2G6_Arg600Gln',\n",
    "#             'PLA2G6_Val310Glu',\n",
    "#             'PLA2G6_Glu567Lys',\n",
    "#             'PLA2G6_Arg538Cys']\n",
    "\n",
    "# variant_list = ['IL2RG_Tyr89Cys', 'KCTD7_Asn273Ile', 'AGXT_Ile279Met',\n",
    "#        'ASNS_Ala6Glu', 'ACSF3_Ser431Tyr', 'HBD_Gly26Asp',\n",
    "#        'SFTPC_Leu188Gln', 'SH3BP2_Gly420Glu', 'DIABLO_Ser126Leu',\n",
    "#        'PLP1_Ser135Phe', 'DES_Ala337Pro', 'SFTPC_Leu194Pro',\n",
    "#        'EIF2B4_Ala228Val', 'TH_Pro121Leu', 'EFHC1_Asp253Tyr',\n",
    "#        'FBP1_Phe194Ser', 'POMT1_Trp560Cys', 'AIPL1_Met79Thr',\n",
    "#        'EFHC1_Cys259Tyr', 'AGXT_Ile202Asn', 'SMAD1_Val3Ala',\n",
    "#        'RP2_Cys86Tyr', 'EFHC1_Met448Thr', 'SFTPC_Ile73Thr',\n",
    "#        'SMN2_Gly287Arg', 'EFHC1_Asp210Asn', 'PKP2_Ala526Thr',\n",
    "#        'RP2_Leu253Arg', 'PLP1_Ile152Thr', 'PMP22_Asp37Val',\n",
    "#        'PLP1_Gly186Cys', 'TAT_Arg433Trp', 'TTPA_Gly246Arg',\n",
    "#        'DCX_Ser129Leu', 'COMP_Asn523Lys', 'FXN_Trp173Gly']\n",
    "\n",
    "# for variant in tqdm(variant_list):\n",
    "#     plot_allele(meta_data, variant, \"GFP\", auroc_df=local_wtvar, site=\"05\", max_intensity=0.999, display=False, imgs_dir=tiff_imgs_dir, output_dir=out_imgs_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc6c095c",
   "metadata": {},
   "source": [
    "## Plot Cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6418cdb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_cells_dir = \"../snakemake_pipeline/outputs/visualize_cells\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "212e2c7b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 7287/7287 [00:05<00:00, 1221.47it/s]\n",
      "100%|██████████| 7665/7665 [00:06<00:00, 1229.72it/s]\n"
     ]
    }
   ],
   "source": [
    "# Filter thresholds\n",
    "min_area_ratio = 0.15\n",
    "max_area_ratio = 0.3\n",
    "min_center = 50\n",
    "max_center = 1030\n",
    "num_mad = 5\n",
    "min_cells = 250\n",
    "\n",
    "batch_profiles = {}\n",
    "for batch_id in batches:\n",
    "    imagecsv_dir = f\"{CP_GALLERY_DIR}/image_csv/cpg0020-varchamp/broad/workspace/assaydev/{batch_id}\"\n",
    "    prof_path = f\"../snakemake_pipeline/outputs/batch_profiles/{batch_id}/profiles.parquet\"\n",
    "    # Get metadata\n",
    "    profiles = pl.scan_parquet(prof_path).select(\n",
    "        [\"Metadata_well_position\", \"Metadata_plate_map_name\", \"Metadata_ImageNumber\", \"Metadata_ObjectNumber\",\n",
    "            \"Metadata_symbol\", \"Metadata_gene_allele\", \"Metadata_control_type\", \"Metadata_Plate\",\n",
    "            \"Nuclei_AreaShape_Area\", \"Cells_AreaShape_Area\", \"Nuclei_AreaShape_Center_X\", \"Nuclei_AreaShape_Center_Y\",\n",
    "            \"Cells_AreaShape_BoundingBoxMaximum_X\", \"Cells_AreaShape_BoundingBoxMaximum_Y\", \"Cells_AreaShape_BoundingBoxMinimum_X\",\n",
    "            \"Cells_AreaShape_BoundingBoxMinimum_Y\",\t\"Cells_AreaShape_Center_X\",\t\"Cells_AreaShape_Center_Y\",\n",
    "            \"Cells_Intensity_MeanIntensity_GFP\", \"Cells_Intensity_MedianIntensity_GFP\", \"Cells_Intensity_IntegratedIntensity_GFP\"],\n",
    "    ).collect()\n",
    "\n",
    "    # Filter based on cell to nucleus area\n",
    "    profiles = profiles.with_columns(\n",
    "                    (pl.col(\"Nuclei_AreaShape_Area\")/pl.col(\"Cells_AreaShape_Area\")).alias(\"Nucleus_Cell_Area\"),\n",
    "                    pl.concat_str([\n",
    "                        \"Metadata_Plate\", \"Metadata_well_position\", \"Metadata_ImageNumber\", \"Metadata_ObjectNumber\",\n",
    "                        ], separator=\"_\").alias(\"Metadata_CellID\"),\n",
    "            ).filter((pl.col(\"Nucleus_Cell_Area\") > min_area_ratio) & (pl.col(\"Nucleus_Cell_Area\") < max_area_ratio))\n",
    "\n",
    "    # Filter cells too close to image edge\n",
    "    profiles = profiles.filter(\n",
    "        ((pl.col(\"Nuclei_AreaShape_Center_X\") > min_center) & (pl.col(\"Nuclei_AreaShape_Center_X\") < max_center) &\n",
    "        (pl.col(\"Nuclei_AreaShape_Center_Y\") > min_center) & (pl.col(\"Nuclei_AreaShape_Center_Y\") < max_center)),\n",
    "    )\n",
    "\n",
    "    # Calculate mean, median and mad of gfp intensity for each allele\n",
    "    ## mean\n",
    "    means = profiles.group_by([\"Metadata_Plate\", \"Metadata_well_position\"]).agg(\n",
    "        pl.col(\"Cells_Intensity_MeanIntensity_GFP\").mean().alias(\"WellIntensityMean\"),\n",
    "    )\n",
    "    profiles = profiles.join(means, on=[\"Metadata_Plate\", \"Metadata_well_position\"])\n",
    "    ## median\n",
    "    medians = profiles.group_by([\"Metadata_Plate\", \"Metadata_well_position\"]).agg(\n",
    "        pl.col(\"Cells_Intensity_MedianIntensity_GFP\").median().alias(\"WellIntensityMedian\"),\n",
    "    )\n",
    "    profiles = profiles.join(medians, on=[\"Metadata_Plate\", \"Metadata_well_position\"])\n",
    "    ## mad\n",
    "    profiles = profiles.with_columns(\n",
    "        (pl.col(\"Cells_Intensity_MedianIntensity_GFP\") - pl.col(\"WellIntensityMedian\")).abs().alias(\"Abs_dev\"),\n",
    "    )\n",
    "    mad = profiles.group_by([\"Metadata_Plate\", \"Metadata_well_position\"]).agg(\n",
    "        pl.col(\"Abs_dev\").median().alias(\"Intensity_MAD\"),\n",
    "    )\n",
    "    profiles = profiles.join(mad, on=[\"Metadata_Plate\", \"Metadata_well_position\"])\n",
    "\n",
    "    # Threshold is 5X\n",
    "    ## Used to be median well intensity + 5*mad implemented by Jess\n",
    "    ## Switching to mean well intensity + 5*mad implemented by Runxi\n",
    "    # profiles = profiles.with_columns(\n",
    "    #     (pl.col(\"WellIntensityMean\") + num_mad*pl.col(\"Intensity_MAD\")).alias(\"Intensity_upper_threshold\"), ## pl.col(\"WellIntensityMedian\")\n",
    "    #     (pl.col(\"WellIntensityMean\") - num_mad*pl.col(\"Intensity_MAD\")).alias(\"Intensity_lower_threshold\"), ## pl.col(\"WellIntensityMedian\")\n",
    "    # )\n",
    "    # ## Filter by intensity MAD\n",
    "    # profiles = profiles.filter(\n",
    "    #     pl.col(\"Cells_Intensity_MeanIntensity_GFP\") <= pl.col(\"Intensity_upper_threshold\"),\n",
    "    # ).filter(\n",
    "    #     pl.col(\"Cells_Intensity_MeanIntensity_GFP\") >= pl.col(\"Intensity_lower_threshold\"),\n",
    "    # )\n",
    "\n",
    "    # Filter out alleles with fewer than 250 cells\n",
    "    # keep_alleles = profiles.group_by(\"Metadata_gene_allele\").count().filter(\n",
    "    #     pl.col(\"count\") >= min_cells,\n",
    "    #     ).select(\"Metadata_gene_allele\").to_series().to_list()\n",
    "    # profiles = profiles.filter(pl.col(\"Metadata_gene_allele\").is_in(keep_alleles))\n",
    "\n",
    "    # add full crop coordinates\n",
    "    profiles = profiles.with_columns(\n",
    "        (pl.col(\"Nuclei_AreaShape_Center_X\") - 50).alias(\"x_low\").round().cast(pl.Int16),\n",
    "        (pl.col(\"Nuclei_AreaShape_Center_X\") + 50).alias(\"x_high\").round().cast(pl.Int16),\n",
    "        (pl.col(\"Nuclei_AreaShape_Center_Y\") - 50).alias(\"y_low\").round().cast(pl.Int16),\n",
    "        (pl.col(\"Nuclei_AreaShape_Center_Y\") + 50).alias(\"y_high\").round().cast(pl.Int16),\n",
    "    )\n",
    "\n",
    "    # Read in all Image.csv to get ImageNumber:SiteNumber mapping and paths\n",
    "    image_dat = []\n",
    "    icf = os.listdir(imagecsv_dir)\n",
    "    for fp in tqdm(icf):\n",
    "        plate, well = fp.split(\"-\")\n",
    "        image_dat.append(pl.read_csv(f\"{imagecsv_dir}/{fp}/Image.csv\").select(\n",
    "            [\n",
    "                \"ImageNumber\",\n",
    "                \"Metadata_Site\",\n",
    "                \"PathName_OrigDNA\",\n",
    "                \"FileName_OrigDNA\",\n",
    "                \"FileName_OrigGFP\",\n",
    "                ],\n",
    "            ).with_columns(\n",
    "            pl.lit(plate).alias(\"Metadata_Plate\"),\n",
    "            pl.lit(well).alias(\"Metadata_well_position\"),\n",
    "            ))\n",
    "    image_dat = pl.concat(image_dat).rename({\"ImageNumber\": \"Metadata_ImageNumber\"})\n",
    "\n",
    "    # Create useful filepaths\n",
    "    image_dat = image_dat.with_columns(\n",
    "        pl.col(\"PathName_OrigDNA\").str.replace(\".*cpg0020-varchamp/\", \"\").alias(\"Path_root\"),\n",
    "    )\n",
    "\n",
    "    image_dat = image_dat.drop([\n",
    "        \"PathName_OrigDNA\",\n",
    "        \"FileName_OrigDNA\",\n",
    "        \"FileName_OrigGFP\",\n",
    "        \"Path_root\",\n",
    "    ])\n",
    "\n",
    "    # Append to profiles\n",
    "    profiles = profiles.join(image_dat, on = [\"Metadata_Plate\", \"Metadata_well_position\", \"Metadata_ImageNumber\"])\n",
    "\n",
    "    # Sort by allele, then image number\n",
    "    profiles = profiles.with_columns(\n",
    "        pl.concat_str([\"Metadata_Plate\", \"Metadata_well_position\", \"Metadata_Site\"], separator=\"_\").alias(\n",
    "            \"Metadata_SiteID\"),\n",
    "        pl.col(\"Metadata_gene_allele\").str.replace(\"_\", \"-\").alias(\"Protein_label\"),\n",
    "    )\n",
    "    profiles = profiles.sort([\"Protein_label\", \"Metadata_SiteID\"])\n",
    "    alleles = profiles.select(\"Protein_label\").to_series().unique().to_list()\n",
    "\n",
    "    batch_profiles[batch_id] = profiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "23e11956",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (20, 36)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>Metadata_well_position</th><th>Metadata_plate_map_name</th><th>Metadata_ImageNumber</th><th>Metadata_ObjectNumber</th><th>Metadata_symbol</th><th>Metadata_gene_allele</th><th>Metadata_control_type</th><th>Metadata_Plate</th><th>Nuclei_AreaShape_Area</th><th>Cells_AreaShape_Area</th><th>Nuclei_AreaShape_Center_X</th><th>Nuclei_AreaShape_Center_Y</th><th>Cells_AreaShape_BoundingBoxMaximum_X</th><th>Cells_AreaShape_BoundingBoxMaximum_Y</th><th>Cells_AreaShape_BoundingBoxMinimum_X</th><th>Cells_AreaShape_BoundingBoxMinimum_Y</th><th>Cells_AreaShape_Center_X</th><th>Cells_AreaShape_Center_Y</th><th>Cells_Intensity_MeanIntensity_GFP</th><th>Cells_Intensity_MedianIntensity_GFP</th><th>Cells_Intensity_IntegratedIntensity_GFP</th><th>Nucleus_Cell_Area</th><th>Metadata_CellID</th><th>WellIntensityMean</th><th>WellIntensityMedian</th><th>Abs_dev</th><th>Intensity_MAD</th><th>Intensity_upper_threshold</th><th>Intensity_lower_threshold</th><th>x_low</th><th>x_high</th><th>y_low</th><th>y_high</th><th>Metadata_Site</th><th>Metadata_SiteID</th><th>Protein_label</th></tr><tr><td>str</td><td>str</td><td>i64</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>i64</td><td>i64</td><td>f64</td><td>f64</td><td>i64</td><td>i64</td><td>i64</td><td>i64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>str</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>i16</td><td>i16</td><td>i16</td><td>i16</td><td>i64</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;A05&quot;</td><td>&quot;B7A1R1_P4&quot;</td><td>41</td><td>11</td><td>&quot;SMAD1&quot;</td><td>&quot;SMAD1&quot;</td><td>&quot;disease_wt&quot;</td><td>&quot;2024_01_19_B7A1R1_P4T1&quot;</td><td>1096</td><td>5814</td><td>81.438869</td><td>415.609489</td><td>168</td><td>458</td><td>58</td><td>359</td><td>96.54128</td><td>413.690918</td><td>0.006644</td><td>0.00633</td><td>38.629356</td><td>0.18851</td><td>&quot;2024_01_19_B7A1R1_P4T1_A05_41_…</td><td>0.004921</td><td>0.003691</td><td>0.002638</td><td>0.001137</td><td>0.010605</td><td>-0.000763</td><td>31</td><td>131</td><td>366</td><td>466</td><td>5</td><td>&quot;2024_01_19_B7A1R1_P4T1_A05_5&quot;</td><td>&quot;SMAD1&quot;</td></tr><tr><td>&quot;A05&quot;</td><td>&quot;B7A1R1_P4&quot;</td><td>41</td><td>24</td><td>&quot;SMAD1&quot;</td><td>&quot;SMAD1&quot;</td><td>&quot;disease_wt&quot;</td><td>&quot;2024_01_19_B7A1R1_P4T1&quot;</td><td>1809</td><td>11405</td><td>88.215589</td><td>704.983416</td><td>153</td><td>743</td><td>0</td><td>633</td><td>71.303639</td><td>693.142744</td><td>0.010516</td><td>0.010121</td><td>119.936142</td><td>0.158615</td><td>&quot;2024_01_19_B7A1R1_P4T1_A05_41_…</td><td>0.004921</td><td>0.003691</td><td>0.006429</td><td>0.001137</td><td>0.010605</td><td>-0.000763</td><td>38</td><td>138</td><td>655</td><td>755</td><td>5</td><td>&quot;2024_01_19_B7A1R1_P4T1_A05_5&quot;</td><td>&quot;SMAD1&quot;</td></tr><tr><td>&quot;A05&quot;</td><td>&quot;B7A1R1_P4&quot;</td><td>41</td><td>18</td><td>&quot;SMAD1&quot;</td><td>&quot;SMAD1&quot;</td><td>&quot;disease_wt&quot;</td><td>&quot;2024_01_19_B7A1R1_P4T1&quot;</td><td>1153</td><td>6510</td><td>881.117953</td><td>631.679965</td><td>943</td><td>668</td><td>842</td><td>533</td><td>891.28894</td><td>614.277266</td><td>0.003122</td><td>0.003021</td><td>20.327037</td><td>0.177112</td><td>&quot;2024_01_19_B7A1R1_P4T1_A05_41_…</td><td>0.004921</td><td>0.003691</td><td>0.00067</td><td>0.001137</td><td>0.010605</td><td>-0.000763</td><td>831</td><td>931</td><td>582</td><td>682</td><td>5</td><td>&quot;2024_01_19_B7A1R1_P4T1_A05_5&quot;</td><td>&quot;SMAD1&quot;</td></tr><tr><td>&quot;A05&quot;</td><td>&quot;B7A1R1_P4&quot;</td><td>41</td><td>28</td><td>&quot;SMAD1&quot;</td><td>&quot;SMAD1&quot;</td><td>&quot;disease_wt&quot;</td><td>&quot;2024_01_19_B7A1R1_P4T1&quot;</td><td>878</td><td>4949</td><td>604.738041</td><td>805.56492</td><td>635</td><td>858</td><td>556</td><td>759</td><td>593.280865</td><td>801.839968</td><td>0.003648</td><td>0.003634</td><td>18.054927</td><td>0.17741</td><td>&quot;2024_01_19_B7A1R1_P4T1_A05_41_…</td><td>0.004921</td><td>0.003691</td><td>0.000057</td><td>0.001137</td><td>0.010605</td><td>-0.000763</td><td>555</td><td>655</td><td>756</td><td>856</td><td>5</td><td>&quot;2024_01_19_B7A1R1_P4T1_A05_5&quot;</td><td>&quot;SMAD1&quot;</td></tr><tr><td>&quot;A05&quot;</td><td>&quot;B7A1R1_P4&quot;</td><td>41</td><td>29</td><td>&quot;SMAD1&quot;</td><td>&quot;SMAD1&quot;</td><td>&quot;disease_wt&quot;</td><td>&quot;2024_01_19_B7A1R1_P4T1&quot;</td><td>877</td><td>3563</td><td>937.348917</td><td>805.080958</td><td>965</td><td>837</td><td>878</td><td>776</td><td>921.106652</td><td>801.920573</td><td>0.004</td><td>0.003816</td><td>14.252918</td><td>0.246141</td><td>&quot;2024_01_19_B7A1R1_P4T1_A05_41_…</td><td>0.004921</td><td>0.003691</td><td>0.000124</td><td>0.001137</td><td>0.010605</td><td>-0.000763</td><td>887</td><td>987</td><td>755</td><td>855</td><td>5</td><td>&quot;2024_01_19_B7A1R1_P4T1_A05_5&quot;</td><td>&quot;SMAD1&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;A05&quot;</td><td>&quot;B7A1R1_P4&quot;</td><td>41</td><td>23</td><td>&quot;SMAD1&quot;</td><td>&quot;SMAD1&quot;</td><td>&quot;disease_wt&quot;</td><td>&quot;2024_01_19_B7A1R1_P4T1&quot;</td><td>1214</td><td>7037</td><td>874.387974</td><td>701.83855</td><td>933</td><td>776</td><td>829</td><td>666</td><td>875.765809</td><td>722.628393</td><td>0.002756</td><td>0.002659</td><td>19.395149</td><td>0.172517</td><td>&quot;2024_01_19_B7A1R1_P4T1_A05_41_…</td><td>0.004921</td><td>0.003691</td><td>0.001033</td><td>0.001137</td><td>0.010605</td><td>-0.000763</td><td>824</td><td>924</td><td>652</td><td>752</td><td>5</td><td>&quot;2024_01_19_B7A1R1_P4T1_A05_5&quot;</td><td>&quot;SMAD1&quot;</td></tr><tr><td>&quot;A05&quot;</td><td>&quot;B7A1R1_P4&quot;</td><td>41</td><td>32</td><td>&quot;SMAD1&quot;</td><td>&quot;SMAD1&quot;</td><td>&quot;disease_wt&quot;</td><td>&quot;2024_01_19_B7A1R1_P4T1&quot;</td><td>1511</td><td>9829</td><td>124.005295</td><td>840.509596</td><td>179</td><td>920</td><td>33</td><td>786</td><td>110.007529</td><td>844.244277</td><td>0.003028</td><td>0.002951</td><td>29.76286</td><td>0.153729</td><td>&quot;2024_01_19_B7A1R1_P4T1_A05_41_…</td><td>0.004921</td><td>0.003691</td><td>0.00074</td><td>0.001137</td><td>0.010605</td><td>-0.000763</td><td>74</td><td>174</td><td>791</td><td>891</td><td>5</td><td>&quot;2024_01_19_B7A1R1_P4T1_A05_5&quot;</td><td>&quot;SMAD1&quot;</td></tr><tr><td>&quot;A05&quot;</td><td>&quot;B7A1R1_P4&quot;</td><td>41</td><td>34</td><td>&quot;SMAD1&quot;</td><td>&quot;SMAD1&quot;</td><td>&quot;disease_wt&quot;</td><td>&quot;2024_01_19_B7A1R1_P4T1&quot;</td><td>1899</td><td>10988</td><td>239.512902</td><td>877.313323</td><td>327</td><td>1033</td><td>103</td><td>800</td><td>206.159629</td><td>916.411631</td><td>0.008434</td><td>0.008457</td><td>92.670951</td><td>0.172825</td><td>&quot;2024_01_19_B7A1R1_P4T1_A05_41_…</td><td>0.004921</td><td>0.003691</td><td>0.004765</td><td>0.001137</td><td>0.010605</td><td>-0.000763</td><td>190</td><td>290</td><td>827</td><td>927</td><td>5</td><td>&quot;2024_01_19_B7A1R1_P4T1_A05_5&quot;</td><td>&quot;SMAD1&quot;</td></tr><tr><td>&quot;A05&quot;</td><td>&quot;B7A1R1_P4&quot;</td><td>41</td><td>8</td><td>&quot;SMAD1&quot;</td><td>&quot;SMAD1&quot;</td><td>&quot;disease_wt&quot;</td><td>&quot;2024_01_19_B7A1R1_P4T1&quot;</td><td>1772</td><td>8106</td><td>757.417043</td><td>333.806433</td><td>860</td><td>417</td><td>716</td><td>268</td><td>768.725265</td><td>327.819023</td><td>0.006469</td><td>0.006435</td><td>52.436602</td><td>0.218604</td><td>&quot;2024_01_19_B7A1R1_P4T1_A05_41_…</td><td>0.004921</td><td>0.003691</td><td>0.002743</td><td>0.001137</td><td>0.010605</td><td>-0.000763</td><td>707</td><td>807</td><td>284</td><td>384</td><td>5</td><td>&quot;2024_01_19_B7A1R1_P4T1_A05_5&quot;</td><td>&quot;SMAD1&quot;</td></tr><tr><td>&quot;A05&quot;</td><td>&quot;B7A1R1_P4&quot;</td><td>41</td><td>33</td><td>&quot;SMAD1&quot;</td><td>&quot;SMAD1&quot;</td><td>&quot;disease_wt&quot;</td><td>&quot;2024_01_19_B7A1R1_P4T1&quot;</td><td>1553</td><td>8916</td><td>452.87508</td><td>873.157759</td><td>564</td><td>952</td><td>387</td><td>806</td><td>459.879991</td><td>872.919919</td><td>0.007502</td><td>0.00755</td><td>66.883713</td><td>0.174181</td><td>&quot;2024_01_19_B7A1R1_P4T1_A05_41_…</td><td>0.004921</td><td>0.003691</td><td>0.003859</td><td>0.001137</td><td>0.010605</td><td>-0.000763</td><td>403</td><td>503</td><td>823</td><td>923</td><td>5</td><td>&quot;2024_01_19_B7A1R1_P4T1_A05_5&quot;</td><td>&quot;SMAD1&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (20, 36)\n",
       "┌────────────┬────────────┬───────────┬───────────┬───┬────────┬───────────┬───────────┬───────────┐\n",
       "│ Metadata_w ┆ Metadata_p ┆ Metadata_ ┆ Metadata_ ┆ … ┆ y_high ┆ Metadata_ ┆ Metadata_ ┆ Protein_l │\n",
       "│ ell_positi ┆ late_map_n ┆ ImageNumb ┆ ObjectNum ┆   ┆ ---    ┆ Site      ┆ SiteID    ┆ abel      │\n",
       "│ on         ┆ ame        ┆ er        ┆ ber       ┆   ┆ i16    ┆ ---       ┆ ---       ┆ ---       │\n",
       "│ ---        ┆ ---        ┆ ---       ┆ ---       ┆   ┆        ┆ i64       ┆ str       ┆ str       │\n",
       "│ str        ┆ str        ┆ i64       ┆ i64       ┆   ┆        ┆           ┆           ┆           │\n",
       "╞════════════╪════════════╪═══════════╪═══════════╪═══╪════════╪═══════════╪═══════════╪═══════════╡\n",
       "│ A05        ┆ B7A1R1_P4  ┆ 41        ┆ 11        ┆ … ┆ 466    ┆ 5         ┆ 2024_01_1 ┆ SMAD1     │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ 9_B7A1R1_ ┆           │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ P4T1_A05_ ┆           │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ 5         ┆           │\n",
       "│ A05        ┆ B7A1R1_P4  ┆ 41        ┆ 24        ┆ … ┆ 755    ┆ 5         ┆ 2024_01_1 ┆ SMAD1     │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ 9_B7A1R1_ ┆           │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ P4T1_A05_ ┆           │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ 5         ┆           │\n",
       "│ A05        ┆ B7A1R1_P4  ┆ 41        ┆ 18        ┆ … ┆ 682    ┆ 5         ┆ 2024_01_1 ┆ SMAD1     │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ 9_B7A1R1_ ┆           │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ P4T1_A05_ ┆           │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ 5         ┆           │\n",
       "│ A05        ┆ B7A1R1_P4  ┆ 41        ┆ 28        ┆ … ┆ 856    ┆ 5         ┆ 2024_01_1 ┆ SMAD1     │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ 9_B7A1R1_ ┆           │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ P4T1_A05_ ┆           │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ 5         ┆           │\n",
       "│ A05        ┆ B7A1R1_P4  ┆ 41        ┆ 29        ┆ … ┆ 855    ┆ 5         ┆ 2024_01_1 ┆ SMAD1     │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ 9_B7A1R1_ ┆           │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ P4T1_A05_ ┆           │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ 5         ┆           │\n",
       "│ …          ┆ …          ┆ …         ┆ …         ┆ … ┆ …      ┆ …         ┆ …         ┆ …         │\n",
       "│ A05        ┆ B7A1R1_P4  ┆ 41        ┆ 23        ┆ … ┆ 752    ┆ 5         ┆ 2024_01_1 ┆ SMAD1     │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ 9_B7A1R1_ ┆           │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ P4T1_A05_ ┆           │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ 5         ┆           │\n",
       "│ A05        ┆ B7A1R1_P4  ┆ 41        ┆ 32        ┆ … ┆ 891    ┆ 5         ┆ 2024_01_1 ┆ SMAD1     │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ 9_B7A1R1_ ┆           │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ P4T1_A05_ ┆           │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ 5         ┆           │\n",
       "│ A05        ┆ B7A1R1_P4  ┆ 41        ┆ 34        ┆ … ┆ 927    ┆ 5         ┆ 2024_01_1 ┆ SMAD1     │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ 9_B7A1R1_ ┆           │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ P4T1_A05_ ┆           │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ 5         ┆           │\n",
       "│ A05        ┆ B7A1R1_P4  ┆ 41        ┆ 8         ┆ … ┆ 384    ┆ 5         ┆ 2024_01_1 ┆ SMAD1     │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ 9_B7A1R1_ ┆           │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ P4T1_A05_ ┆           │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ 5         ┆           │\n",
       "│ A05        ┆ B7A1R1_P4  ┆ 41        ┆ 33        ┆ … ┆ 923    ┆ 5         ┆ 2024_01_1 ┆ SMAD1     │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ 9_B7A1R1_ ┆           │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ P4T1_A05_ ┆           │\n",
       "│            ┆            ┆           ┆           ┆   ┆        ┆           ┆ 5         ┆           │\n",
       "└────────────┴────────────┴───────────┴───────────┴───┴────────┴───────────┴───────────┴───────────┘"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def crop_allele(allele: str, profile_df: pl.DataFrame, meta_plate: str, rep: str, site: str=\"5\") -> None:\n",
    "    \"\"\"Crop images and save metadata as numpy arrays for one allele.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    allele : String\n",
    "        Name of allele to process\n",
    "    profile_df : String\n",
    "        Dataframe with pathname and cell coordinates\n",
    "    img_dir : String\n",
    "        Directory where all images are stored\n",
    "    out_dir : String\n",
    "        Directory where numpy arrays should be saved\n",
    "    \"\"\"\n",
    "    allele_df = profile_df.filter(\n",
    "        (pl.col(\"Metadata_gene_allele\")==allele) &\n",
    "        (pl.col(\"Metadata_Site\")==int(site)) &\n",
    "        (pl.col(\"Metadata_plate_map_name\")==meta_plate) &\n",
    "        (pl.col(\"Metadata_Plate\").str.contains(rep))\n",
    "    )\n",
    "    return allele_df\n",
    "\n",
    "allele_df = crop_allele(\"SMAD1\", batch_profiles[\"2024_01_23_Batch_7\"], \"B7A1R1_P4\", \"T1\", site=\"5\")\n",
    "allele_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f4f748ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Image for SMAD1_Val3Ala already exists.\n",
      "7 SMAD1 A05\n"
     ]
    }
   ],
   "source": [
    "# Compute distances from edges and find the most centered well\n",
    "def compute_distance_cell(row, col, edge=1080):\n",
    "    return min(row - 1, edge - row, col - 1, edge - col)  # Distance from nearest edge\n",
    "    \n",
    "\n",
    "def plot_allele_cell(pm, variant, sel_channel, auroc_df, site=\"05\", max_intensity=0.99, display=False, imgs_dir=tiff_imgs_dir, output_dir=out_imgs_dir):\n",
    "    cmap = channel_to_cmap(sel_channel)\n",
    "    channel = channel_dict[sel_channel]\n",
    "    auroc = auroc_df.filter(pl.col(\"allele_0\")==variant)[\"AUROC_mean\"].mean()\n",
    "\n",
    "    if os.path.exists(os.path.join(output_dir, f\"{variant}_{sel_channel}_{auroc:.3f}_cells.png\")):\n",
    "        print(f\"Image for {variant} already exists.\")\n",
    "        # return None\n",
    "\n",
    "    ## get the number of wells/images per allele\n",
    "    wt = variant.split(\"_\")[0]\n",
    "    wt_wells = pm.filter(pl.col(\"gene_allele\") == wt).select(\"imaging_well\").to_pandas().values.flatten()\n",
    "    var_wells = pm.filter(pl.col(\"gene_allele\") == variant).select(\"imaging_well\").to_pandas().values.flatten()\n",
    "    plate_map = pm.filter(pl.col(\"gene_allele\") == variant).select(\"plate_map_name\").to_pandas().values.flatten()\n",
    "\n",
    "    if len(wt_wells) > 1:\n",
    "        # Get coordinates of wells\n",
    "        well_coords = [well_to_coordinates(w) for w in set([ref_well_pl for ref_well_pl in wt_wells])]\n",
    "        # Sort wells by max distance from edges (descending)\n",
    "        wt_wells = [max(well_coords, key=lambda x: compute_distance(x[1], x[2]))[0]]\n",
    "\n",
    "    pm_var = pm.filter((pl.col(\"imaging_well\").is_in(np.concatenate([wt_wells, var_wells])))&(pl.col(\"plate_map_name\").is_in(plate_map))).sort(\"control_type\")\n",
    "    # print(pm_var)\n",
    "    plt.clf()\n",
    "\n",
    "    fig, axes = plt.subplots((len(wt_wells)+len(var_wells))*2, 4, figsize=(15, 16), sharex=True, sharey=True)\n",
    "    # fig2, axes2 = plt.subplots((len(wt_wells)+len(var_wells))*2, 4, figsize=(15, 16), sharex=True, sharey=True)\n",
    "    for wt_var, pm_row in enumerate(pm_var.iter_rows(named=True)):\n",
    "        # print(pm_row)\n",
    "        if pm_row[\"control_type\"] == \"allele\":\n",
    "            well = var_wells[0]\n",
    "            allele = variant\n",
    "        else:\n",
    "            well = wt_wells[0]\n",
    "            allele = wt\n",
    "\n",
    "        for i in range(8):\n",
    "            plot_idx = i+wt_var*4*2\n",
    "            if i < 4:\n",
    "                sel_plate = pm_row[\"imaging_plate_R1\"]\n",
    "            else:\n",
    "                sel_plate = pm_row[\"imaging_plate_R2\"]\n",
    "                \n",
    "            batch = batch_dict[sel_plate.split(\"_\")[0]]\n",
    "            batch_img_dir = f'{imgs_dir}/{batch}/images'\n",
    "            letter = well[0]\n",
    "            row = letter_dict[letter]\n",
    "            col = well[1:3]\n",
    "            \n",
    "            # print(i, allele, well)\n",
    "            plate_img_dir = plate_dict[sel_plate][f\"T{i%4+1}\"]\n",
    "            img_file = f\"r{row}c{col}f{site}p01-ch{channel}sk1fk1fl1.tiff\"\n",
    "            img = imread(f\"{batch_img_dir}/{plate_img_dir}/Images/{img_file}\", as_gray=True)\n",
    "\n",
    "            ## Full images\n",
    "            # axes2.flatten()[plot_idx].imshow(img, vmin=0, vmax=np.percentile(img, max_intensity*100), cmap=cmap)\n",
    "            # plot_label = f\"{sel_channel}:{sel_plate},T{i%4+1}\\nWell:{well},Site:{site}\\n{allele}\"\n",
    "            # axes2.flatten()[plot_idx].text(0.03, 0.97, plot_label, color='white', fontsize=10,\n",
    "            #         verticalalignment='top', horizontalalignment='left', transform=axes2.flatten()[plot_idx].transAxes,\n",
    "            #         bbox=dict(facecolor='black', alpha=0.3, linewidth=2))\n",
    "\n",
    "            ## Draw cells\n",
    "            cell_allele_coord_df = crop_allele(allele, batch_profiles[batch], sel_plate, f\"T{i%4+1}\", site=site[-1]) # \"2024_01_19_B7A1R1_P4T1\"\n",
    "            cell_allele_coord_df = cell_allele_coord_df.with_columns(\n",
    "                pl.struct(\"Cells_AreaShape_Center_X\", \"Cells_AreaShape_Center_Y\") # 'Nuclei_AreaShape_Center_X', 'Nuclei_AreaShape_Center_Y'\n",
    "                .map_elements(lambda x: compute_distance_cell(x['Cells_AreaShape_Center_X'], x['Cells_AreaShape_Center_Y']), return_dtype=pl.Float32).cast(pl.Int16)\n",
    "                .alias('dist2edge')\n",
    "            ).sort(\"dist2edge\", descending=True)\n",
    "            \n",
    "            if cell_allele_coord_df.is_empty():\n",
    "                axes.flatten()[plot_idx].set_visible(False)\n",
    "                print(i, allele, well)\n",
    "                continue\n",
    "            else:\n",
    "                x, y = int(cell_allele_coord_df[\"Cells_AreaShape_Center_X\"].to_numpy()[0]), int(cell_allele_coord_df[\"Cells_AreaShape_Center_Y\"].to_numpy()[0])\n",
    "                # print(x, y)\n",
    "                img_sub = img[\n",
    "                    y-128:y+128, x-128:x+128\n",
    "                ]\n",
    "                # print(img_sub.shape)\n",
    "                if img_sub.shape[0] == 0 or img_sub.shape[1] == 0:\n",
    "                    axes.flatten()[plot_idx].set_visible(False)\n",
    "                    continue\n",
    "\n",
    "                # print(img_sub.shape)\n",
    "                # Get key intensity percentiles\n",
    "                p99 = np.percentile(img_sub, 99)\n",
    "                p100 = np.max(img_sub)\n",
    "\n",
    "                # If there's a big jump between 99th and max, lower the threshold\n",
    "                if p100 > 1.5 * p99:\n",
    "                    # print(\"More contrast\")\n",
    "                    max_intensity = 0.95  # More contrast\n",
    "                else:\n",
    "                    max_intensity = max_intensity  # Allow more of the bright tail\n",
    "\n",
    "                axes.flatten()[plot_idx].imshow(img_sub, vmin=0, vmax=np.percentile(img_sub, max_intensity*100), cmap=cmap)  ## np.percentile(img_sub, max_intensity*100)\n",
    "                plot_label = f\"{sel_channel}:{sel_plate},T{i%4+1}\\nWell:{well},Site:{site}\\n{allele}\"\n",
    "                axes.flatten()[plot_idx].text(0.03, 0.97, plot_label, color='white', fontsize=10,\n",
    "                        verticalalignment='top', horizontalalignment='left', transform=axes.flatten()[plot_idx].transAxes,\n",
    "                        bbox=dict(facecolor='black', alpha=0.3, linewidth=2))\n",
    "                int_95 = str(int(round(np.percentile(img_sub, 95))))\n",
    "                axes.flatten()[plot_idx].text(0.95, 0.05, f\"95th Percentile\\nIntensity:{int_95}\", color='white', fontsize=12,\n",
    "                            verticalalignment='bottom', horizontalalignment='right', transform=axes.flatten()[plot_idx].transAxes,\n",
    "                            bbox=dict(facecolor='black', alpha=0.3, linewidth=2))\n",
    "                axes.flatten()[plot_idx].axis(\"off\")\n",
    "                \n",
    "                # x_min, y_min, x_max, y_max = int(cell_allele_coord_df[\"Cells_AreaShape_BoundingBoxMinimum_X\"].to_numpy()[0]), \\\n",
    "                #                              int(cell_allele_coord_df[\"Cells_AreaShape_BoundingBoxMinimum_Y\"].to_numpy()[0]), \\\n",
    "                #                              int(cell_allele_coord_df[\"Cells_AreaShape_BoundingBoxMaximum_X\"].to_numpy()[0]), \\\n",
    "                #                              int(cell_allele_coord_df[\"Cells_AreaShape_BoundingBoxMaximum_Y\"].to_numpy()[0])\n",
    "                # print(x_min, y_min, x_max, y_max)\n",
    "            #     for cell_idx, cell_row in enumerate(cell_allele_coord_df.head().iter_rows(named=True)):\n",
    "            #         x_nuc, y_nuc = int(cell_row[\"Nuclei_AreaShape_Center_X\"]), int(cell_row[\"Nuclei_AreaShape_Center_Y\"])\n",
    "            #         # Create a Rectangle patch\n",
    "            #         rect = patches.Rectangle((x_nuc-5, y_nuc-5), 10, 10, linewidth=2, edgecolor='red', facecolor='none')\n",
    "            #         axes2.flatten()[plot_idx].add_patch(rect)\n",
    "            #         rect = patches.Rectangle((x-128, y-128), 256, 256, linewidth=2, edgecolor='red', facecolor='none')\n",
    "            #         axes2.flatten()[plot_idx].add_patch(rect)\n",
    "            #         # x_min, y_min, x_max, y_max = int(cell_row[\"Cells_AreaShape_BoundingBoxMinimum_X\"]), int(cell_row[\"Cells_AreaShape_BoundingBoxMinimum_Y\"]), \\\n",
    "            #         #                              int(cell_row[\"Cells_AreaShape_BoundingBoxMaximum_X\"]), int(cell_row[\"Cells_AreaShape_BoundingBoxMaximum_Y\"])\n",
    "            #         # print(x_min, y_min, x_max, y_max)\n",
    "            #         # rect = patches.Rectangle((x_min, y_min), x_max-x_min, y_max-y_min, linewidth=2, edgecolor='red', facecolor='none')\n",
    "            #         # Add the patch to the Axes\n",
    "            #         axes2.flatten()[plot_idx].add_patch(rect)\n",
    "            #         break\n",
    "\n",
    "            # int_95 = str(int(round(np.percentile(img, 95))))\n",
    "            # axes2.flatten()[plot_idx].axhline(y=img.shape[1]//2, color='red', linestyle='--', linewidth=1)\n",
    "            # axes2.flatten()[plot_idx].axvline(x=img.shape[0]//2, color='red', linestyle='--', linewidth=1)\n",
    "            # axes2.flatten()[plot_idx].text(0.95, 0.05, f\"95th Percentile\\nIntensity:{int_95}\", color='white', fontsize=12,\n",
    "            #                verticalalignment='bottom', horizontalalignment='right', transform=axes2.flatten()[plot_idx].transAxes)\n",
    "            # axes2.flatten()[plot_idx].axis(\"off\")\n",
    "            \n",
    "    \n",
    "    fig.tight_layout()\n",
    "    fig.subplots_adjust(wspace=.01, hspace=-0.2, top=.99)\n",
    "\n",
    "    # fig2.tight_layout()\n",
    "    # fig2.subplots_adjust(wspace=.01, hspace=-0.2, top=.99)\n",
    "    \n",
    "    if display:\n",
    "        plt.show()\n",
    "\n",
    "    if output_dir:\n",
    "        fig.savefig(os.path.join(output_dir, f\"{variant}_{sel_channel}_{auroc:.3f}_cells.png\"), dpi=400, bbox_inches='tight')\n",
    "        plt.close(fig)\n",
    "\n",
    "plot_allele_cell(meta_data, \"SMAD1_Val3Ala\", \"GFP\", auroc_df=local_wtvar, site=\"05\", max_intensity=0.95, display=False, imgs_dir=tiff_imgs_dir, output_dir=out_cells_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ae8e3e77",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot_allele_cell(meta_data, \"PLA2G6_Val310Glu\", \"GFP\", auroc_df=local_wtvar, site=\"05\", max_intensity=0.99, display=True, imgs_dir=tiff_imgs_dir, output_dir=out_cells_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8fc0cb40",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/6 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Image for PLA2G6_Thr585Ile already exists.\n",
      "5 PLA2G6_Thr585Ile N13\n",
      "5 PLA2G6 D13\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 17%|█▋        | 1/6 [00:04<00:20,  4.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Image for PLA2G6_Leu481Gln already exists.\n",
      "3 PLA2G6_Leu481Gln P13\n",
      "5 PLA2G6_Leu481Gln P13\n",
      "7 PLA2G6_Leu481Gln P13\n",
      "5 PLA2G6 D13\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 33%|███▎      | 2/6 [00:07<00:15,  3.88s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Image for PLA2G6_Arg600Gln already exists.\n",
      "5 PLA2G6_Arg600Gln J13\n",
      "5 PLA2G6 D13\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 50%|█████     | 3/6 [00:11<00:11,  3.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Image for PLA2G6_Val310Glu already exists.\n",
      "3 PLA2G6_Val310Glu B15\n",
      "5 PLA2G6_Val310Glu B15\n",
      "6 PLA2G6_Val310Glu B15\n",
      "5 PLA2G6 D13\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 67%|██████▋   | 4/6 [00:15<00:07,  3.85s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Image for PLA2G6_Glu567Lys already exists.\n",
      "5 PLA2G6_Glu567Lys F13\n",
      "7 PLA2G6_Glu567Lys F13\n",
      "5 PLA2G6 D13\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 83%|████████▎ | 5/6 [00:19<00:03,  3.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Image for PLA2G6_Arg538Cys already exists.\n",
      "3 PLA2G6_Arg538Cys H13\n",
      "5 PLA2G6_Arg538Cys H13\n",
      "5 PLA2G6 D13\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 6/6 [00:23<00:00,  3.89s/it]\n"
     ]
    }
   ],
   "source": [
    "# variant_list = ['RP2_Cys86Tyr', 'RP2_Leu253Arg', 'HBD_Gly26Asp', 'EFHC1_Asp253Tyr',\n",
    "#        'PKP2_Ala526Thr', 'PMP22_Asp37Val', 'SMAD1_Val3Ala',\n",
    "#        'DCX_Ser129Leu', 'TTPA_Gly246Arg', 'FXN_Trp173Gly',\n",
    "#        'EFHC1_Cys259Tyr', 'IL2RG_Tyr89Cys', 'SFTPC_Ile73Thr',\n",
    "#        'PLP1_Gly186Cys', 'ASNS_Ala6Glu', 'EFHC1_Met448Thr',\n",
    "#        'SH3BP2_Gly420Glu', 'KCTD7_Asn273Ile', 'EIF2B4_Ala228Val',\n",
    "#        'ACSF3_Ser431Tyr', 'SFTPC_Leu188Gln', 'SMN2_Gly287Arg',\n",
    "#        'POMT1_Trp560Cys', 'SFTPC_Leu194Pro', 'PLP1_Ile152Thr',\n",
    "#        'EFHC1_Asp210Asn', 'FBP1_Phe194Ser', 'TAT_Arg433Trp',\n",
    "#        'DES_Ala337Pro', 'COMP_Asn523Lys', 'AIPL1_Met79Thr',\n",
    "#        'PLP1_Ser135Phe', 'DIABLO_Ser126Leu', 'AGXT_Ile202Asn']\n",
    "\n",
    "# variant_list_2 = ['ACSF3_Met266Val',\n",
    "#  'ACSF3_Gly119Asp',\n",
    "#  'ACSF3_Arg558Trp',\n",
    "#  'ACSF3_Arg471Trp',\n",
    "#  'ACSF3_Pro243Leu',\n",
    "#  'ACSF3_Asp457Asn',\n",
    "#  'ACSF3_Thr358Ile']\n",
    "\n",
    "variant_list_3 = ['PLA2G6_Thr585Ile',\n",
    "            'PLA2G6_Leu481Gln',\n",
    "            'PLA2G6_Arg600Gln',\n",
    "            'PLA2G6_Val310Glu',\n",
    "            'PLA2G6_Glu567Lys',\n",
    "            'PLA2G6_Arg538Cys']\n",
    "\n",
    "# variant_list += variant_list_2 + variant_list_3\n",
    "\n",
    "for variant in tqdm(variant_list_3):\n",
    "    plot_allele_cell(meta_data, variant, \"GFP\", auroc_df=local_wtvar, site=\"05\", max_intensity=0.99, display=False, imgs_dir=tiff_imgs_dir, output_dir=out_cells_dir)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "varchamp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
